---
title: "20240310 DHS grid search"
output: html_document
author: Max Dippel
date: "2023-10-03"
---

Load packages
```{r,eval=FALSE}
library(data.table)
library(tidyverse)
library(RColorBrewer)
library(rmarkdown)
library(GenomicRanges)
library(ggplot2)
library(viridis)
library(pals)
```

Establish directories
```{r directories, eval=FALSE}
# This is the directory for the data exist and the tables will be printed
data.dir <- "~/Desktop/Ho et al. writings/Code_from_scratch2/data/"
# This is the directory for the DHS data you need to run this analysis. Import DHS data in hg19. DHS data is found at https://zenodo.org/records/3838751 
dhs.dir <- "~/Desktop/Ho et al. writings/Code_from_scratch2/data/hg19_dhs/"
# This is the output directory for the plots generated in this markdown
plots.dir <- "~/Desktop/Ho et al. writings/Code_from_scratch2/plots"
```

Create dhs.merge.jurkat and dhs.merge.cd4 to use in the analysis. These are intermediate files for each analysis. Currently set to create jurkat data. You can modify for CD4 or any other cell type to get the files you need. 
```{r dhs merge, eval=FALSE}
# Created MPRA data using big_table.Rmd code. 
mpra.dir <- paste0(data.dir,"20241111_tcell_mpra_merge_all.txt")
mpra <- read.table(mpra.dir, sep="\t", header=T)
# Give MPRA a strand column
mpra$strand<-"+"

# Import DHS data in hg19. DHS data is found at https://zenodo.org/records/3838751
dhs.pos<- read.delim(gzfile(paste0(dhs.dir,"DHS_Index_and_Vocabulary_hg19_WM20190703.txt.gz")), header=T, stringsAsFactors = F)
sample.dat<-read.delim(paste0(dhs.dir,"DHS_Index_and_Vocabulary_metadata.tsv"), sep="\t", header=T, stringsAsFactors = F)
dhs.dat<-fread(paste0(dhs.dir,"dat_bin_FDR01_hg19.txt.gz"))

# Format mpra data
  # subset data
  mpra<-subset(mpra, project=="TGWAS", select=c(chr, snp_start, snp_end, mpra_sig))
  # Create strand column
  mpra$strand<-"+"
  # Make into a Granges file
  mpra.se<-makeGRangesFromDataFrame(mpra,  seqnames = "chr", start.field = "snp_start", end.field = "snp_end",keep.extra.columns=TRUE)
  
# format dhs.pos data
  # subset data
  dhs.pos<-subset(dhs.pos, select=c(seqname, start, end)) 
  # Make a peak name for easier indexing
  dhs.pos$peak<-paste0(dhs.pos$seqname, ":", dhs.pos$start, "_", dhs.pos$end) 
  # make a granges object out of the dataframe
  dhs.pos.se<-makeGRangesFromDataFrame(dhs.pos, seqnames = "seqname", start.field = "start", end.field = "end",keep.extra.columns=TRUE) #Make into GRanges format
  
# Find DHS peaks that overlap with MPRA
  #Find DHS peaks that overlap with MPRA SNPs
  overlap_peaks<-data.frame(subsetByOverlaps(dhs.pos.se, mpra.se, type="any"), stringsAsFactors = F)$peak 
  #Extract row indices of overlapping peaks from DHS file
  overlap_index<-which(dhs.pos$peak%in%overlap_peaks) 
  #Subset DHS peaks file by index of overlapping peaks
  dhs.pos<-dhs.pos[overlap_index,] 
  # remove the previous objects
  rm(mpra.se)
  rm(dhs.pos.se)
  
  dhs.dat<-dhs.dat[overlap_index,] #Subset DHS matrix file by index of overlapping peaks
  dhs.dat<-cbind(dhs.pos[,c(1:3)],dhs.dat)
  
# Format DHS sample information
  sample.dat<-subset(sample.dat, !is.na(library.order),select=c(library.order,Biosample.name, System,Subsystem,  Organ, Biosample.type, Biological.state))

# Merge all these data sets
  # For each cell type, merge DHS peaks from that cell type. Add a column for each cell type, and list it's peak membership    for any sample in that cell type
  dhs.merge.dat<-dhs.dat[,c(1:3)]
  for(c in unique(sample.dat$Biosample.name)){
  libraries<-paste0("V", subset(sample.dat, Biosample.name==c)$library.order)
  if(length(libraries)>1){
    dhs.merge.dat$newcol<-apply(dhs.dat[,c(libraries)],1, max)
  }else{
    dhs.merge.dat$newcol<-dhs.dat[,c(libraries)]
  }
  names(dhs.merge.dat)<-gsub("newcol", c, names(dhs.merge.dat))
  }
  
# Cell type specific line. 
dhs.merge.tcell <- subset(dhs.merge.dat, select=c("seqname","start","end","CD4"))
# Cell type specific line. 
write.table(dhs.merge.tcell,paste0(data.dir,"dhs.merge.tcell.txt"), row.names=F, col.names=T, sep="\t", quote=F)

# Cell type specific line. 
dhs.merge.jurkat <- subset(dhs.merge.dat, select=c("seqname","start","end","Jurkat"))
# Cell type specific line. 
write.table(dhs.merge.jurkat,paste0(data.dir,"dhs.merge.jurkat.txt"), row.names=F, col.names=T, sep="\t", quote=F)

```


#####################
# Primary T-cell
#####################

Load DHS data
```{r, eval=FALSE}
# Import DHS data in hg19. DHS data is found at https://zenodo.org/records/3838751
dhs.pos<- read.delim(gzfile(paste0(dhs.dir,"DHS_Index_and_Vocabulary_hg19_WM20190703.txt.gz")), header=T, stringsAsFactors = F)
# format dhs.pos data
  # subset data
  dhs.pos<-subset(dhs.pos, select=c(seqname, start, end)) 
  # Make a peak name for easier indexing
  dhs.pos$peak<-paste0(dhs.pos$seqname, ":", dhs.pos$start, "_", dhs.pos$end) 
  # make a granges object out of the dataframe
  dhs.pos.se<-makeGRangesFromDataFrame(dhs.pos, seqnames = "seqname", start.field = "start", end.field = "end",keep.extra.columns=TRUE) #Make into GRanges format
  
# This is a table I created by subsetting the original dhs.merged.dat to only include CD4 T-cells. This can be done with other cell types when I need to do this analysis for other cell types
dhs.merge.dat <- read.table(paste0(data.dir,"dhs.merge.tcell.txt"), sep="\t", header=T)
```

Load MPRA data and Tcell data
```{r, eval=FALSE}
mpra.dir <- paste0(data.dir,"20241111_tcell_mpra_merge_filtered.txt")
mpra <- read.table(mpra.dir, sep="\t", header=T)
# Give MPRA a strand column
mpra$strand<-"+"
```

This runs a loop to find the max fold changes.
```{r max FC column, eval=FALSE}
#For each MPRA variant, find the fold change of the allele with the largest absolute fold change
# This creates a column full of NAs for the loop to fill
mpra$max_abs_fc <- NA
# This loop calculates the column of max_abs_fc. It takes the two alleles A and B and calculates which one's absolute value is larger. That larger value is the value of the new column (can be positive or negative). 
for(i in 1:nrow(mpra)){
  if(abs(mpra[i,]$A.log2FC)>abs(mpra[i,]$B.log2FC)){
    mpra[i,]$max_abs_fc<-mpra[i,]$A.log2FC
  }else{
    mpra[i,]$max_abs_fc<-mpra[i,]$B.log2FC
  }
}
```

This runs a loop to find the max fold change p values with the Benjamini Hochberg correction.
```{r max BH column, eval=FALSE}
#For each MPRA variant, find the fold change of the allele with the largest absolute fold change
# This creates a column full of NAs for the loop to fill
mpra$max.logPadj_BH<-NA
# This loop calculates the column of max_abs_fc. It takes the two alleles A and B and calculates which one's absolute value is larger. That larger value is the value of the new column (can be positive or negative). 
for(i in 1:nrow(mpra)){
  if(abs(mpra[i,]$A.logPadj_BH)>abs(mpra[i,]$B.logPadj_BH)){
    mpra[i,]$max.logPadj_BH<-mpra[i,]$A.logPadj_BH
  }else{
    mpra[i,]$max.logPadj_BH<-mpra[i,]$B.logPadj_BH
  }
}
```

This runs a loop to find the max fold change p values with the bonferroni correction.
```{r max BF column}
#For each MPRA variant, find the fold change of the allele with the largest absolute fold change
# This creates a column full of NAs for the loop to fill
mpra$max.logPadj_BF<-NA
# This loop calculates the column of max_abs_fc. It takes the two alleles A and B and calculates which one's absolute value is larger. That larger value is the value of the new column (can be positive or negative). 
for(i in 1:nrow(mpra)){
  if(abs(mpra[i,]$A.logPadj_BF)>abs(mpra[i,]$B.logPadj_BF)){
    mpra[i,]$max.logPadj_BF<-mpra[i,]$A.logPadj_BF
  }else{
    mpra[i,]$max.logPadj_BF<-mpra[i,]$B.logPadj_BF
  }
}
```


# Loop for Tcell DHS enrichemnt and p-value  
```{r Tcell loop}
#skew_seq <-c(0, 0.25, 0.5, 0.75, 1, 1.25, 1.5, 1.75, 2, 3)
FC_seq <- c(0, 0.25, 0.5, 0.75, 1, 1.25, 1.5, 1.75, 2, 3)
AdjP_seq <- c(0, 0.25, 0.5, 0.75, 1, 1.25, 1.5, 1.75, 2, 3)
#FC_seq <- c(0.25, 1)
#AdjP_seq <- c(0.25, 1)

tcell.precision.skew.dat <- NULL
for(j in AdjP_seq){
  for(k in FC_seq){
mpra2 <- mpra
   
#Manually temporarily set these FDRs and adjusted p-values that are NA to -1 for easier TRUE/FALSE testing
mpra2$Skew.logFDR[is.na(mpra2$Skew.logFDR)]<- (-1)
mpra2$max.logPadj_BH[is.na(mpra2$max.logPadj_BH)]<- (-1)

#For each variant, assign whether it has no activity (nEnhancer_nSkew), is a pCRE only (Enhancer_nSkew) or emVAR (Enhancer_Skew)
mpra2$mpra_sig<-"nEnhancer_nSkew"
for(i in 1:nrow(mpra2)){
  if(mpra2[i,]$Skew.logFDR>1 & (mpra2[i,]$max.logPadj_BH>j & mpra[i,]$max_abs_fc>k) ){
    mpra2[i,]$mpra_sig<-"Enhancer_Skew"
  }else if(mpra2[i,]$Skew.logFDR<=1 & (mpra2[i,]$max.logPadj_BH>j & mpra[i,]$max_abs_fc>k) ){
    mpra2[i,]$mpra_sig<-"Enhancer_nSkew"
  }
}
#Set those -1 FDRs and p-values back to NA
mpra2$Skew.logFDR[mpra2$Skew.logFDR<0]<- NA
mpra2$max.logPadj_BH[mpra2$max.logPadj_BH<0]<- NA
   
#Make into a Granges file
mpra.se<-makeGRangesFromDataFrame(mpra2,  seqnames = "chr", start.field = "snp_start", end.field = "snp_end",keep.extra.columns=TRUE)

# This code finds where the genomic positions of the MPRA data and DHS peak data overlap
#Find DHS peaks that overlap with MPRA
overlap_peaks<-data.frame(subsetByOverlaps(dhs.pos.se, mpra.se, type="any"), stringsAsFactors = F)$peak #Find DHS peaks that overlap with MPRA SNPs
overlap_index<-which(dhs.pos$peak%in%overlap_peaks) #Extract row indices of overlapping peaks from DHS file
dhs.pos<-dhs.pos[overlap_index,] #Subset DHS peaks file by index of overlapping peaks
# This removes the mpra.se and dhs.pos.se objects

# Generate peak indices for MPRA SNPs for faster downstream processing
# This creates a new dataframe with seqname, start and end 
dhs.pos<-dhs.merge.dat[,c(1:3)]
# This creates an index column which are just sequentially numbered (1,2,3 etc.)
dhs.pos$index<-seq(1:nrow(dhs.pos))
# This adds the index column to the last data set as well
dhs.merge.dat$index<-seq(1:nrow(dhs.merge.dat))
# This creates a column called peak_index to be filled in by the loop
mpra2$peak_index<-NA

# Create a function to find the peak index for a single row
find_peak_index <- function(row) {
  dhs_subset <- subset(dhs.pos, seqname == row$chr & start <= row$snp_end & end >= row$snp_end)
  if (nrow(dhs_subset) > 0) {
    return(dhs_subset$index[1])
  } else {
    return(NA)  # No matching DHS peak found
  }
}

# Apply the function to each row of mpra
mpra2$peak_index <- sapply(1:nrow(mpra2), function(i) find_peak_index(mpra2[i, ]))

# This creates a new dataframe which is the mpra data, but it is subset to only to pCREs and emVars and observations with a peak_index value
pcre_index<-subset(mpra2, mpra_sig%in%c("Enhancer_nSkew", "Enhancer_Skew") & !is.na(peak_index))$peak_index #peak indices for MPRA pCRE SNPs only
# This subsets the other observations
non_pcre_index<-subset(mpra2, mpra_sig=="nEnhancer_nSkew" & !is.na(peak_index))$peak_index #peak indices for MPRA non-pCRE SNPs

# This code creates the sample dataframe which in this case is just CD4. It is originally made with all the biosample names. 
sample.dat <- data.frame(c("CD4"))
colnames(sample.dat) <- "sample"

# Calculate enrichments of MPRA pCREs in DHS peaks for each cell type

# This created an empty dataframe to be filled by the loop
dat.plot<-data.frame(sample= sample.dat$sample, pcre_fold=0, pcre_p=0, a=0,b=0,c=0,d=0,odds=0,lower.conf=0,upper.conf=0, stringsAsFactors = F)
# This loop creates fold enrichment and p-values for the pCREs
for(i in 1:nrow(dat.plot)){
  
  #Find the column in the dhs data that correspond to that cell type
  cell_col<-which(names(dhs.merge.dat)==dat.plot[i,]$sample)
  
  #find which peak_indexes for the peaks in that cell type
  peak_index<-dhs.merge.dat[dhs.merge.dat[,cell_col]==1,]$index

  #Calculate enrichment
  a<-length(intersect(peak_index, pcre_index)) #Number of pCRE in DHS peak
  b<-length(intersect(peak_index, non_pcre_index)) #Number of non-pCRE in DHS peak
  c<-nrow(subset(mpra2, mpra_sig%in%c("Enhancer_nSkew", "Enhancer_Skew")))-a #Number of pCRE not in DHS
  d<-nrow(subset(mpra2, mpra_sig=="nEnhancer_nSkew"))-b #Number of non-pCRE not in DHS
  dat.plot[i,]$pcre_fold<-(a/(a+c))/(b/(b+d)) #Calcualte fold enrichment
  dat.plot[i,]$pcre_p<-fisher.test(rbind(c(a,b), c(c, d)), alternative="two.sided")$p.value #Calculate p-value
  dat.plot[i,]$a <- a
  dat.plot[i,]$b <- b
  dat.plot[i,]$c <- c
  dat.plot[i,]$d <- d
  dat.plot[i,]$odds<-fisher.test(rbind(c(a,b), c(c, d)))$estimate #Calculate odds ratio
  dat.plot[i,]$lower.conf<-fisher.test(rbind(c(a,b), c(c, d)))$conf.int[1] #Calculate lower confidence interval
  dat.plot[i,]$upper.conf<-fisher.test(rbind(c(a,b), c(c, d)))$conf.int[2] #Calculate higher confidence interval
}
pcre_p <- dat.plot$pcre_p
pcre_enrichment <- dat.plot$pcre_fold
TP_sample_size <- dat.plot$a
FP_sample_size <- dat.plot$b
FN_sample_size <- dat.plot$c
TN_sample_size <- dat.plot$d
odds_ratio <- dat.plot$odds
lower_conf <- dat.plot$lower.conf
upper_conf <- dat.plot$upper.conf

   tcell.precision.skew.dat <- rbind(tcell.precision.skew.dat, data.frame(k, j, pcre_p,pcre_enrichment,TP_sample_size,FP_sample_size,FN_sample_size,TN_sample_size,odds_ratio,lower_conf,upper_conf)) # makes the data frame used for the heat map
  }
}

names(tcell.precision.skew.dat) <- c("FC", "pAdj", "pcre_p", "pcre_fold","TP_sample_size","FP_sample_size","FN_sample_size","TN_sample_size","odds_ratio","lower_conf","upper_conf")
```

Save the T-cell precision skew data. 
```{r}
write.table(tcell.precision.skew.dat,paste0(data.dir,"tcell.precision.skew.dat.txt"), row.names=F, col.names=T, sep="\t", quote=F)
```

#####################
# Unstimulated Jurkat
#####################
Load DHS data
```{r}
# Import DHS data in hg19. DHS data is found at https://zenodo.org/records/3838751
dhs.pos<- read.delim(gzfile(paste0(dhs.dir,"DHS_Index_and_Vocabulary_hg19_WM20190703.txt.gz")), header=T, stringsAsFactors = F)
# format dhs.pos data
  # subset data
  dhs.pos<-subset(dhs.pos, select=c(seqname, start, end)) 
  # Make a peak name for easier indexing
  dhs.pos$peak<-paste0(dhs.pos$seqname, ":", dhs.pos$start, "_", dhs.pos$end) 
  # make a granges object out of the dataframe
  dhs.pos.se<-makeGRangesFromDataFrame(dhs.pos, seqnames = "seqname", start.field = "start", end.field = "end",keep.extra.columns=TRUE) #Make into GRanges format
  
# This is a table I created by subsetting the original dhs.merged.dat to only include jurkat cells. This can be done with other cell types when I need to do this analysis for other cell types
dhs.merge.dat <- read.table(paste0(data.dir,"dhs.merge.jurkat.txt"), sep="\t", header=T)
```

Load MPRA data and Tcell data
```{r}
mpra.dir <- paste0(data.dir,"20241111_jurkat_mpra_merge_filtered.txt")
mpra <- read.table(mpra.dir, sep="\t", header=T)
# Give MPRA a strand column
mpra$strand<-"+"
```

This runs a loop to find the max fold changes.
```{r max FC column}
#For each MPRA variant, find the fold change of the allele with the largest absolute fold change
# This creates a column full of NAs for the loop to fill
mpra$max_abs_fc <- NA
# This loop calculates the column of max_abs_fc. It takes the two alleles A and B and calculates which one's absolute value is larger. That larger value is the value of the new column (can be positive or negative). 
for(i in 1:nrow(mpra)){
  if(abs(mpra[i,]$A.log2FC)>abs(mpra[i,]$B.log2FC)){
    mpra[i,]$max_abs_fc<-mpra[i,]$A.log2FC
  }else{
    mpra[i,]$max_abs_fc<-mpra[i,]$B.log2FC
  }
}
```

This runs a loop to find the max fold change p values with the Benjamini Hochberg correction.
```{r max BH column}
#For each MPRA variant, find the fold change of the allele with the largest absolute fold change
# This creates a column full of NAs for the loop to fill
mpra$max.logPadj_BH<-NA
# This loop calculates the column of max_abs_fc. It takes the two alleles A and B and calculates which one's absolute value is larger. That larger value is the value of the new column (can be positive or negative). 
for(i in 1:nrow(mpra)){
  if(abs(mpra[i,]$A.logPadj_BH)>abs(mpra[i,]$B.logPadj_BH)){
    mpra[i,]$max.logPadj_BH<-mpra[i,]$A.logPadj_BH
  }else{
    mpra[i,]$max.logPadj_BH<-mpra[i,]$B.logPadj_BH
  }
}
```

This runs a loop to find the max fold change p values with the bonferroni correction.
```{r max BF column}
#For each MPRA variant, find the fold change of the allele with the largest absolute fold change
# This creates a column full of NAs for the loop to fill
mpra$max.logPadj_BF<-NA
# This loop calculates the column of max_abs_fc. It takes the two alleles A and B and calculates which one's absolute value is larger. That larger value is the value of the new column (can be positive or negative). 
for(i in 1:nrow(mpra)){
  if(abs(mpra[i,]$A.logPadj_BF)>abs(mpra[i,]$B.logPadj_BF)){
    mpra[i,]$max.logPadj_BF<-mpra[i,]$A.logPadj_BF
  }else{
    mpra[i,]$max.logPadj_BF<-mpra[i,]$B.logPadj_BF
  }
}
```

# Loop for Unstim Jurkat DHS enrichemnt and p-value  
```{r Tcell loop}
#skew_seq <-c(0, 0.25, 0.5, 0.75, 1, 1.25, 1.5, 1.75, 2, 3)
FC_seq <- c(0, 0.25, 0.5, 0.75, 1, 1.25, 1.5, 1.75, 2, 3)
AdjP_seq <- c(0, 0.25, 0.5, 0.75, 1, 1.25, 1.5, 1.75, 2, 3)
#FC_seq <- c(0.25, 1)
# AdjP_seq <- c(0.25, 1)

unstim.jurkat.precision.skew.dat <- NULL
for(j in AdjP_seq){
  for(k in FC_seq){
mpra2 <- mpra
   
#Manually temporarily set these FDRs and adjusted p-values that are NA to -1 for easier TRUE/FALSE testing
mpra2$Skew.logFDR[is.na(mpra2$Skew.logFDR)]<- (-1)
mpra2$max.logPadj_BH[is.na(mpra2$max.logPadj_BH)]<- (-1)

#For each variant, assign whether it has no activity (nEnhancer_nSkew), is a pCRE only (Enhancer_nSkew) or emVAR (Enhancer_Skew)
mpra2$mpra_sig<-"nEnhancer_nSkew"
for(i in 1:nrow(mpra2)){
  if(mpra2[i,]$Skew.logFDR>1 & (mpra2[i,]$max.logPadj_BH>j & mpra[i,]$max_abs_fc>k) ){
    mpra2[i,]$mpra_sig<-"Enhancer_Skew"
  }else if(mpra2[i,]$Skew.logFDR<=1 & (mpra2[i,]$max.logPadj_BH>j & mpra[i,]$max_abs_fc>k) ){
    mpra2[i,]$mpra_sig<-"Enhancer_nSkew"
  }
}
#Set those -1 FDRs and p-values back to NA
mpra2$Skew.logFDR[mpra2$Skew.logFDR<0]<- NA
mpra2$max.logPadj_BH[mpra2$max.logPadj_BH<0]<- NA
   
#Make into a Granges file
mpra.se<-makeGRangesFromDataFrame(mpra2,  seqnames = "chr", start.field = "snp_start", end.field = "snp_end",keep.extra.columns=TRUE)

# This code finds where the genomic positions of the MPRA data and DHS peak data overlap
#Find DHS peaks that overlap with MPRA
overlap_peaks<-data.frame(subsetByOverlaps(dhs.pos.se, mpra.se, type="any"), stringsAsFactors = F)$peak #Find DHS peaks that overlap with MPRA SNPs
overlap_index<-which(dhs.pos$peak%in%overlap_peaks) #Extract row indices of overlapping peaks from DHS file
dhs.pos<-dhs.pos[overlap_index,] #Subset DHS peaks file by index of overlapping peaks
# This removes the mpra.se and dhs.pos.se objects

# Generate peak indices for MPRA SNPs for faster downstream processing
# This creates a new dataframe with seqname, start and end 
dhs.pos<-dhs.merge.dat[,c(1:3)]
# This creates an index column which are just sequentially numbered (1,2,3 etc.)
dhs.pos$index<-seq(1:nrow(dhs.pos))
# This adds the index column to the last data set as well
dhs.merge.dat$index<-seq(1:nrow(dhs.merge.dat))
# This creates a column called peak_index to be filled in by the loop
mpra2$peak_index<-NA

# Create a function to find the peak index for a single row
find_peak_index <- function(row) {
  dhs_subset <- subset(dhs.pos, seqname == row$chr & start <= row$snp_end & end >= row$snp_end)
  if (nrow(dhs_subset) > 0) {
    return(dhs_subset$index[1])
  } else {
    return(NA)  # No matching DHS peak found
  }
}

# Apply the function to each row of mpra
mpra2$peak_index <- sapply(1:nrow(mpra2), function(i) find_peak_index(mpra2[i, ]))

# This creates a new dataframe which is the mpra data, but it is subset to only to pCREs and emVars and observations with a peak_index value
pcre_index<-subset(mpra2, mpra_sig%in%c("Enhancer_nSkew", "Enhancer_Skew") & !is.na(peak_index))$peak_index #peak indices for MPRA pCRE SNPs only
# This subsets the other observations
non_pcre_index<-subset(mpra2, mpra_sig=="nEnhancer_nSkew" & !is.na(peak_index))$peak_index #peak indices for MPRA non-pCRE SNPs

# This code creates the sample dataframe which in this case is just CD4. It is originally made with all the biosample names. 
sample.dat <- data.frame(c("Jurkat"))
colnames(sample.dat) <- "sample"

# Calculate enrichments of MPRA pCREs in DHS peaks for each cell type

# This created an empty dataframe to be filled by the loop
dat.plot<-data.frame(sample= sample.dat$sample, pcre_fold=0, pcre_p=0, a=0,b=0,c=0,d=0,odds=0,lower.conf=0,upper.conf=0, stringsAsFactors = F)
# This loop creates fold enrichment and p-values for the pCREs
for(i in 1:nrow(dat.plot)){
  
  #Find the column in the dhs data that correspond to that cell type
  cell_col<-which(names(dhs.merge.dat)==dat.plot[i,]$sample)
  
  #find which peak_indexes for the peaks in that cell type
  peak_index<-dhs.merge.dat[dhs.merge.dat[,cell_col]==1,]$index

  #Calculate enrichment
  a<-length(intersect(peak_index, pcre_index)) #Number of pCRE in DHS peak
  b<-length(intersect(peak_index, non_pcre_index)) #Number of non-pCRE in DHS peak
  c<-nrow(subset(mpra2, mpra_sig%in%c("Enhancer_nSkew", "Enhancer_Skew")))-a #Number of pCRE not in DHS
  d<-nrow(subset(mpra2, mpra_sig=="nEnhancer_nSkew"))-b #Number of non-pCRE not in DHS
  dat.plot[i,]$pcre_fold<-(a/(a+c))/(b/(b+d)) #Calcualte fold enrichment
  dat.plot[i,]$pcre_p<-fisher.test(rbind(c(a,b), c(c, d)), alternative="two.sided")$p.value #Calculate p-value
  dat.plot[i,]$a <- a
  dat.plot[i,]$b <- b
  dat.plot[i,]$c <- c
  dat.plot[i,]$d <- d
  dat.plot[i,]$odds<-fisher.test(rbind(c(a,b), c(c, d)))$estimate #Calculate odds ratio
  dat.plot[i,]$lower.conf<-fisher.test(rbind(c(a,b), c(c, d)))$conf.int[1] #Calculate lower confidence interval
  dat.plot[i,]$upper.conf<-fisher.test(rbind(c(a,b), c(c, d)))$conf.int[2] #Calculate higher confidence interval
}
pcre_p <- dat.plot$pcre_p
pcre_enrichment <- dat.plot$pcre_fold
TP_sample_size <- dat.plot$a
FP_sample_size <- dat.plot$b
FN_sample_size <- dat.plot$c
TN_sample_size <- dat.plot$d
odds_ratio <- dat.plot$odds
lower_conf <- dat.plot$lower.conf
upper_conf <- dat.plot$upper.conf

   unstim.jurkat.precision.skew.dat <- rbind(unstim.jurkat.precision.skew.dat, data.frame(k, j, pcre_p,pcre_enrichment,TP_sample_size,FP_sample_size,FN_sample_size,TN_sample_size,odds_ratio,lower_conf,upper_conf)) # makes the data frame used for the heat map
  }
}

names(unstim.jurkat.precision.skew.dat) <- c("FC", "pAdj", "pcre_p", "pcre_fold","TP_sample_size","FP_sample_size","FN_sample_size","TN_sample_size","odds_ratio","lower_conf","upper_conf")
```

Save the jurkat precision skew data. Unhash to save. 
```{r, eval=FALSE}
write.table(unstim.jurkat.precision.skew.dat,paste0(data.dir,"unstim.jurkat.precision.skew.dat.txt"), row.names=F, col.names=T, sep="\t", quote=F)
```

Plot the Primary T-cell grid search data 
```{r}
tcell.precision.skew.dat <- read.table(paste0(data.dir,"tcell.precision.skew.dat.txt"), sep="\t", header=T)

FC_AdjP_tcell_DHS_p_enrich_new_heat_map <- ggplot(data.frame(tcell.precision.skew.dat), aes(x = as.character(pAdj), y = as.character(FC))) +
  geom_point(aes(size = TP_sample_size, color=pcre_fold)) + 
  xlab("MPRA expression FDR p-value") + ylab("MPRA Expression") +  
   scale_size_continuous(range = c(2, 6),limits = c(0, 450)) +
   scale_color_gradientn(limits = c(0, 32),colors=kovesi.rainbow_bgyrm_35_85_c71(7),breaks=c(0,2,4,6,8,16,32),values=c(0,0.0625,0.125,0.1875,0.25,0.5,1),na.value = "black") + 
  labs(title="Primary Tcell active elements enrichment for DHS sites", subtitle="Each dot is the results of DHS enrichment for a different classification of active elements",size="Number of active elements in DHS sites", color= "Enrichment for DHS sites" ) +
  scale_x_discrete(guide = guide_axis(n.dodge=2)) + theme_bw() + theme(plot.background = element_blank(),panel.background = element_blank(),panel.grid.major = element_blank(), panel.grid.minor = element_blank(),panel.border = element_blank())
FC_AdjP_tcell_DHS_p_enrich_new_heat_map 

# save plot
ggsave("FC_AdjP_tcell_DHS_p_enrich_heat_map_rainbow.pdf",plot = FC_AdjP_tcell_DHS_p_enrich_new_heat_map, path = plots.dir, bg='#ffffff',height=5,width=10)
```

Plot the jurkat grid search data
```{r}
unstim.jurkat.precision.skew.dat <- read.table(paste0(data.dir,"unstim.jurkat.precision.skew.dat.txt"), sep="\t", header=T)

FC_AdjP_unstim.jurkat_DHS_p_enrich_new_heat_map <- ggplot(data.frame(unstim.jurkat.precision.skew.dat), aes(x = as.character(pAdj), y = as.character(FC))) +
  geom_point(aes(size = TP_sample_size, color=pcre_fold)) + 
  xlab("MPRA expression FDR p-value") + ylab("MPRA Expression") +  
   scale_size_continuous(range = c(2, 6),limits = c(0, 450)) +
   scale_color_gradientn(limits = c(0, 32),colors=kovesi.rainbow_bgyrm_35_85_c71(7),breaks=c(0,2,4,6,8,16,32),values=c(0,0.0625,0.125,0.1875,0.25,0.5,1),na.value = "black") + 
  labs(title="Jurkat active elements enrichment for DHS sites", subtitle="Each dot is the results of DHS enrichment for a different classification of active elements",size="Number of active elements in DHS sites", color= "Enrichment for DHS sites" ) +
  scale_x_discrete(guide = guide_axis(n.dodge=2)) + theme_bw() + theme(plot.background = element_blank(),panel.background = element_blank(),panel.grid.major = element_blank(), panel.grid.minor = element_blank(),panel.border = element_blank())
FC_AdjP_unstim.jurkat_DHS_p_enrich_new_heat_map 

# save plot
ggsave("FC_AdjP_unstim.jurkat_DHS_p_enrich_heat_map_rainbow.pdf",plot = FC_AdjP_unstim.jurkat_DHS_p_enrich_new_heat_map, path = plots.dir, bg='#ffffff',height=5,width=10)
```


Now that you have the grid search you are more equipped to create the variant category filter. 


Assign variant category filters to each table (This code is also in the MPRA big table code, but I am putting it here if you would like to change the filteres based on the grid search results): 

## Variant Category Filter Primary T-cells

Create a filter to classify mpra varaints as emVars, pCREs and no activity variants. This is the Primary T-cell filter which additonally specifies log2FC of enhancer activity > 1 for at least one allele per element. 
```{r, warning=FALSE}
mpra.name <- "20241111_tcell"
filter <- "tcell"
# There are three parameters which determine the primary t-cell filter as currently constructed
Skew_FDR_log_p_value <- 1
FC_BF_p_value <- 2
FC_value <- 1

mpra_list <- c("_mpra_merge_all.txt","_mpra_merge_filtered.txt","_mpra_merge_hg38_all.txt","_mpra_merge_hg38_filtered.txt")
##########
for(M in mpra_list){
mpra <- read.table(paste0(data.dir,mpra.name,M), sep="\t", header=T)
######
if(filter=="tcell"){
#
# Create absolute value columns 
mpra$abs.A.log2FC <- abs(mpra$A.log2FC)
mpra$abs.B.log2FC <- abs(mpra$B.log2FC)
# Reset the mpra_sig column
mpra$mpra_sig<- NA
# For every row in the mpra, if the absolute value of the fold change is above "FC_value" and the BH p-vlaue is above "FC_BF_p_value" for either allele and the LogSkew FDR is above "Skew_FDR_log_p_value", then the variant is labeled as an emVar (Enhancer_Skew). Else, if the absolute value of the fold change is above "FC_value" and the BH p-value is above "FC_BF_p_value" for either allele, then it is labeled as a pCRE (Enhancer_nSkew). If it is niether of those it is a no activity variant (nEnhancer_nSkew). 
###
for(i in 1:nrow(mpra)){
  mpra[i,]$mpra_sig<-ifelse(((mpra[i,]$abs.A.log2FC>FC_value &mpra[i,]$A.logPadj_BF>FC_BF_p_value) | (mpra[i,]$abs.B.log2FC>FC_value & mpra[i,]$B.logPadj_BF>FC_BF_p_value)) & mpra[i,]$Skew.logFDR>=Skew_FDR_log_p_value, "Enhancer_Skew", ifelse((mpra[i,]$abs.A.log2FC>FC_value &mpra[i,]$A.logPadj_BF>FC_BF_p_value) | (mpra[i,]$abs.B.log2FC>FC_value & mpra[i,]$B.logPadj_BF>FC_BF_p_value), "Enhancer_nSkew", "nEnhancer_nSkew"))
}
###
mpra<- subset(mpra, select=-c(abs.A.log2FC,abs.B.log2FC))
}
######
write.table(mpra, (paste0(data.dir,mpra.name,M)), row.names=F, col.names=T, sep="\t", quote=F)
}
##########
```

## Variant Category Filter Jurkat

Create a filter to classify mpra varaints as emVars (high activity and differential expression between alleles), pCREs (high activity) and no activity variants. This is the unstimulated jurkat filter because it does not have a log2FC cut-off. It is slightly different to the filter in Mouri et al. 2022 to be consistant with the primary T-cell filter. 
```{r, eval=FALSE}
mpra.name <- "20241111_jurkat"
filter <- "jurkat"
# There are two parameters which determine the jurkat filter as currently constructed
Skew_FDR_log_p_value <- 1
FC_BF_p_value <- 2

# MPRA's in order: MPRA hg19 with all variants, MPRA hg19 with variants filtered for >20 plasmids, MPRA hg38 with all variants, MPRA hg38 with variants filtered for >20 plasmids
mpra_list <- c("_mpra_merge_all.txt","_mpra_merge_filtered.txt","_mpra_merge_hg38_all.txt","_mpra_merge_hg38_filtered.txt")
##########
for(M in mpra_list){
# MPRA hg19 with all variants
mpra <- read.table(paste0(data.dir,mpra.name,M), sep="\t", header=T)
######
if(filter=="jurkat"){
mpra$mpra_sig<- NA
# For every row in the mpra, if the BH p-vlaue is above "FC_BF_p_value" for either allele and the LogSkew FDR is above "Skew_FDR_log_p_value", then the variant is labeled as an emVar (Enhancer_Skew). Else, if the BH p-vlaue is above "FC_BF_p_value" for either allele, then it is labeled as a pCRE (Enhancer_nSkew). If it is neither of those it is a no activity variant (nEnhancer_nSkew). 
###
for(i in 1:nrow(mpra)){
  mpra[i,]$mpra_sig<-ifelse(((mpra[i,]$A.logPadj_BF>FC_BF_p_value) | (mpra[i,]$B.logPadj_BF>FC_BF_p_value)) & mpra[i,]$Skew.logFDR>=Skew_FDR_log_p_value, "Enhancer_Skew", ifelse((mpra[i,]$A.logPadj_BF>FC_BF_p_value) | (mpra[i,]$B.logPadj_BF>FC_BF_p_value), "Enhancer_nSkew", "nEnhancer_nSkew"))
}
###
}
######
write.table(mpra, (paste0(data.dir,mpra.name,M)), row.names=F, col.names=T, sep="\t", quote=F)
}
##########

```









