---
title: "MPRA plots"
output: html_document
author: Max Dippel
date: "2024-10-28"
---

The goal of this code is to create plots for the MPRA data. The plots will be Figure 1 b-f, supplementary figures 3 & 4. There will also be some data generated in this code that will end up as supplementary tables 3, 4, 8, and 34. 

Load the packages
```{r packages, message=FALSE, warning=FALSE}
library(readxl)
library(data.table)
library(stringr)
library(IRanges)
#library(SummarizedExperiment)
library(ggplot2)
library(pheatmap)
library(ChIPseeker)
library(clusterProfiler)
library(reshape2)
library(filelock)
#library(GenomicFeatures)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
library(SNPlocs.Hsapiens.dbSNP144.GRCh37)
library(BSgenome.Hsapiens.UCSC.hg19)
library(RColorBrewer)
library(ggrepel)
library(R.utils)
library(pegas)
library(gt)
library(ggvenn)
library(GenomicRanges)
library(motifbreakR)
```

Establish directories
```{r}
# plots.dir is the place where your plots are stored 
plots.dir <- "~/Desktop/Ho et al. writings/Code_from_scratch2/plots/"
# data.dir is the place where your data are stored 
data.dir <- "~/Desktop/Ho et al. writings/Code_from_scratch2/data/"
# This is the directory where you put the original MPRA analysis code created by Mike Guo. It is located here.  https://zenodo.org/records/6302248
mhguo.dir <- "~/Desktop/Ho et al. writings/Code_from_scratch2/data/mhguo1-T_cell_MPRA-5c36361/"
```

Set up MPRA data
```{r}
# The main MPRA you are using to make all the plot
mpra.name1 <- "20241111_tcell"
# The main MPRA secondary MPRA you are using to compare to the first. This is only really necessary for the venn diagram and the comparison scatter plot so you can skip this if you don't have two MPRAs to compare. 
mpra.name2 <- "20241111_jurkat"

# make sure you can actually load the MPRA data. Use the big table code to create the tables that I want you to load here. 
mpra1<-read.delim(paste0(data.dir,mpra.name1,"_mpra_merge_filtered.txt"), header=T, stringsAsFactors = F, sep="\t")
mpra2<-read.delim(paste0(data.dir,mpra.name2,"_mpra_merge_filtered.txt"), header=T, stringsAsFactors = F, sep="\t")

# Give a name to each MPRA merge file to go into the title of each graph which is relevant. 
mpra1_name <- "Primary Tcell"
mpra2_name<- "Jurkat"

# Colors for each MPRA in the plots. 
color1 <- "#E14A4A"
color2 <- "#5D93CD"

# What is the stimulation status of the cells? We going to eliminate stimulated (S) or unstimulated (U) cells.
eliminate_stim_status <- "U" # Stimulated so eliminate unstimulated cells with U
#eliminate_stim_status <- "S" <- # Unstimulated so eliminate stimulated cells with S
```

# Venn Diagram
```{r venn diagram}
# Title for the venn diagram
title <- "Variants with high activity and differential expression between alleles (emVars)"

# Create a dataframe of the emVar, pCRE and no activity variant counts for all three data sets
df <- data.frame(cbind(
  c(nrow(subset(mpra1, mpra_sig == "Enhancer_Skew")),
    nrow(subset(mpra1, mpra_sig == "Enhancer_nSkew")),
    nrow(subset(mpra1, mpra_sig == "nEnhancer_nSkew"))),
  c(nrow(subset(mpra2, mpra_sig == "Enhancer_Skew")),
    nrow(subset(mpra2, mpra_sig == "Enhancer_nSkew")),
    nrow(subset(mpra2, mpra_sig == "nEnhancer_nSkew")))),
  row.names = c("emVars","pCREs","No Activity"))
# Give the data frame the proper headings
  colnames(df) <- c(mpra1_name,mpra2_name)
  gtable_compare <- gt(df, rownames_to_stub = TRUE)
  gtable_compare
  
set1 <- subset(mpra1, mpra_sig == "Enhancer_Skew", select=c(SNP))
set1 <- paste(set1$SNP)
set2 <- subset(mpra2, mpra_sig == "Enhancer_Skew", select=c(SNP))
set2 <- paste(set2$SNP)
  
# Creating a list for the venn diagram
B <-list(mpra1_name=set1, mpra2_name=set2)
names(B) <- c(mpra1_name,mpra2_name)

venn_diagram_plot <- ggvenn(B, set_name_size = 3, fill_color = c(color1,color2)) + 
  ggtitle(title) + 
  theme(plot.title = element_text(hjust = 0.5, size=14, face="bold"))
venn_diagram_plot 

# creating the plot name from the mpra.name1
  plots_name <- paste0("emvars_ven_diagram_",mpra.name1,"_",mpra.name2,".pdf")
# Saving the plot in the plots directory 
  ggsave(plots_name,plot = venn_diagram_plot, path = plots.dir)


# Hypergeometric statistical test for overlap 
  # https://zhouhufeng.wordpress.com/2014/04/21/statistically-test-of-venn-diagram/
  # Hypergeometric test for how many variants were expected vs. observed to overlap between jurkat and primary t cell MPRA emVars
  tcell_emvars <- subset(mpra1, select=c(SNP,mpra_sig))
  names(tcell_emvars) <- c("SNP","tcell_sig")
  jurkat_emvars <- subset(mpra2, select=c(SNP,mpra_sig))         
  names(jurkat_emvars) <- c("SNP","jurkat_sig")
  hyper.test.dat <- merge(tcell_emvars,jurkat_emvars, by="SNP")
  q <- nrow(subset(hyper.test.dat, tcell_sig=="Enhancer_Skew" & jurkat_sig=="Enhancer_Skew"))
  m <- nrow(subset(hyper.test.dat, tcell_sig=="Enhancer_Skew"))
  n <- nrow(hyper.test.dat) - nrow(subset(hyper.test.dat, tcell_sig=="Enhancer_Skew"))
  k <- nrow(subset(hyper.test.dat, jurkat_sig=="Enhancer_Skew"))
  phyper(q,m,n,k, lower.tail = FALSE, log.p = FALSE)
  
# Calculate the expected overlap 
Expected_variants <- (nrow(subset(hyper.test.dat, tcell_sig=="Enhancer_Skew"))/nrow(hyper.test.dat)) * (nrow(subset(hyper.test.dat, jurkat_sig=="Enhancer_Skew"))/nrow(hyper.test.dat)) * nrow(hyper.test.dat)
print(paste0("The observed number of overlapped variants is ", nrow(subset(hyper.test.dat, tcell_sig=="Enhancer_Skew" & jurkat_sig=="Enhancer_Skew")), " and the expected number of variants is ", Expected_variants))
```

Allelic Bias scatter plot
```{r}
# Load the MPRAs

mpra1$skew <- ifelse(mpra1$mpra_sig == "Enhancer_Skew", "emVar", "no_skew")
mpra1 <- subset(mpra1, select=c(SNP,LogSkew,Skew.logFDR,skew))
names(mpra1) <- c("SNP","mpra1_LogSkew","mpra1_Skew.logFDR","mpra1_skew")
mpra2$skew <- ifelse(mpra2$mpra_sig == "Enhancer_Skew", "emVar", "no_skew")
mpra2 <- subset(mpra2, select=c(SNP,LogSkew,Skew.logFDR,skew))
names(mpra2) <- c("SNP","mpra2_LogSkew","mpra2_Skew.logFDR","mpra2_skew")
# Merge the MPRAs
double_mpra_scatter <- merge(mpra1, mpra2, by="SNP")

# Allelic Bias plot #6da934 #e14ae1
Allelic_bias_scatter_plot <- ggplot(data=double_mpra_scatter, aes(x=mpra1_LogSkew, y=mpra2_LogSkew))+
  geom_point(data=subset(double_mpra_scatter, mpra2_skew!="emVar" & mpra1_skew!="emVar"),color="gray90",size=2) + 
  geom_point(data=subset(double_mpra_scatter, mpra2_skew=="emVar"),color="#E14A4A",size=2) + 
  geom_point(data=subset(double_mpra_scatter, mpra1_skew=="emVar"),color="#5D93CD",size=2) + 
  geom_point(data=subset(double_mpra_scatter, mpra1_skew=="emVar" & mpra2_skew=="emVar"),color="#6da934",size=2) + 
  labs(title =paste0(mpra1_name, " and ", mpra2_name," MPRA Allelic Bias Comparsion"), x ="Primary Tcell Allelic Bias", y = "Jurkat Allelic Bias") +
  theme_bw() 
Allelic_bias_scatter_plot


plots_name <- paste0(mpra.name1,"_",mpra.name2,"_allelic_bias_scatter_plot",".pdf")
# Saving the plot in the plots directory 
ggsave(plots_name,plot = Allelic_bias_scatter_plot, path = plots.dir, width = 10, height = 6)

# Do a Pearson's coreelation terst to see if there is a correlation. 
cor.test(x=double_mpra_scatter$mpra1_LogSkew,y=double_mpra_scatter$mpra2_LogSkew, alternative = c("two.sided"), method = c("pearson"), conf.level = 0.95)
```


#######################

# hg19

########################

Load MPRA
```{r mpra hg19}
mpra_merge <- read.table(paste0(data.dir,mpra.name1,"_mpra_merge_filtered.txt"), sep="\t", header=T)
```


# Volcano plot
```{r volcano}
# Change name
mpra <- mpra_merge

# Subset MPRA
mpra<-subset(mpra, project=="TGWAS" & !is.na(LogSkew),select=c(SNP, A.log2FC, B.log2FC, LogSkew,mpra_sig))
  
#For each MPRA variant, find the fold change of the allele with the largest absolute fold change
mpra$max_abs_fc<-NA

for(i in 1:nrow(mpra)){
  if(abs(mpra[i,]$A.log2FC)>abs(mpra[i,]$B.log2FC)){
    mpra[i,]$max_abs_fc<-mpra[i,]$A.log2FC
  }else{
    mpra[i,]$max_abs_fc<-mpra[i,]$B.log2FC
  }
}

#Create temporary dataframe with all MPRA alleles
mpra.all<-mpra
mpra.all$mpra_sig_expanded<-"all"

#Create temporary dataframe with only pCRE alleles
mpra.pcre<-subset(mpra, mpra_sig%in%c("Enhancer_nSkew","Enhancer_Skew"))
mpra.pcre$mpra_sig_expanded<-"pcre"

#Create temporary dataframe with only emVars
mpra.emvar<-subset(mpra, mpra_sig=="Enhancer_Skew")
mpra.emvar$mpra_sig_expanded<-"emvar"

#Combined temporary dataframes
mpra.expanded<-rbind(mpra.all, mpra.pcre, mpra.emvar)

#Generate volcano plot 
mpra_volcano_plot <- ggplot() + 
  geom_point(data=mpra,aes(x=max_abs_fc, y=LogSkew), alpha=0.5, pch=16, color="grey50", size=1) +
  geom_point(data=mpra[mpra$mpra_sig!="nEnhancer_nSkew",],aes(x=max_abs_fc, y=LogSkew), alpha=0.2, pch=16, color="goldenrod", size=0.5) +
  geom_point(data=mpra[mpra$mpra_sig=="Enhancer_Skew",],aes(x=max_abs_fc, y=LogSkew), alpha=1, pch=16, color="firebrick4", size=1) +
  theme_bw()+ylab("MPRA Log2(allelic skew)")+xlab("Expression log2(max_allele)")+
  coord_cartesian() + ggtitle("MPRA volcano plot", subtitle=mpra1_name)
mpra_volcano_plot


# Make variant category a factor in upcoming plots
mpra.expanded$mpra_sig_expanded <- factor(mpra.expanded$mpra_sig_expanded, levels=c("all", "pcre", "emvar"))

#Generate density plot for MPRA signal (absolute fold change)
MPRA_signal_density_plot <- ggplot(mpra.expanded, aes(max_abs_fc, fill=mpra_sig_expanded))+geom_density(alpha=0.4, col="white")+ scale_fill_manual(values=c("grey50","goldenrod", "firebrick4"))+theme_bw()+coord_cartesian(xlim=c(-2,7))+ggtitle("MPRA signal density", subtitle=mpra1_name)+xlab("maximum absolute fold change")
MPRA_signal_density_plot

#Generate density plot for MPRA allelic skew
MPRA_allelic_skew_density_plot <- ggplot(mpra.expanded, aes(LogSkew, fill=mpra_sig_expanded))+geom_density(alpha=0.4, col="white")+scale_fill_manual(values=c("grey50","goldenrod", "firebrick4"))+theme_bw()+coord_cartesian(xlim=c(-2,2))+ggtitle("MPRA Allelic Skew", subtitle=mpra1_name)
MPRA_allelic_skew_density_plot

  # creating the plot name from the mpra.name1
plots_name1 <- paste0(mpra.name1,"_Volcano",".pdf")
# Saving the plot in the plots directory 
ggsave(plots_name1,plot = mpra_volcano_plot, path = plots.dir, width = 6, height = 7)

# creating the plot name from the mpra.name1
plots_name2 <- paste0(mpra.name1,"_signal_density",".pdf")
# Saving the plot in the plots directory 
ggsave(plots_name2,plot = MPRA_signal_density_plot, path = plots.dir,width = 7, height = 7)

# creating the plot name from the mpra.name1
plots_name3 <- paste0(mpra.name1,"_allelic_skew_density",".pdf")
# Saving the plot in the plots directory 
ggsave(plots_name3,plot = MPRA_allelic_skew_density_plot, path = plots.dir, width = 7, height = 7)
```

# Variant category table
```{r emvars table}
# Change name
mpra <- mpra_merge
# Create table of variant categories
df <- data.frame(cbind(
  c(nrow(subset(mpra, mpra_sig == "Enhancer_Skew")),
    nrow(subset(mpra, mpra_sig == "Enhancer_nSkew")),
    nrow(subset(mpra, mpra_sig == "nEnhancer_nSkew")))),
  row.names = c("emVars","pCREs only","No Activity"))
# Give the data frame the proper headings
  colnames(df) <- c(paste0(mpra1_name, "MPRA"))
  gtable_compare <- gt(df, rownames_to_stub = TRUE)
  gtable_compare
  
# Answer some important question about the MPRA  
  
# How many active elements are there?
# Create absolute value columns 
mpra$abs.A.log2FC <- abs(mpra$A.log2FC)
mpra$abs.B.log2FC <- abs(mpra$B.log2FC)
# Total pCREs (including emVars) with activity filter 
nrow(subset(mpra, ((abs.A.log2FC>1 & A.logPadj_BF>2) | (abs.B.log2FC>1 & B.logPadj_BF>2))))
# Total pCREs (including emVars) without activity filter (active elements)
nrow(subset(mpra, ((A.logPadj_BF>2) | (B.logPadj_BF>2))))

# What was the maximum number of emvars linked to a single lead variant and what was the lead variant?
emvars_only <- subset(mpra, mpra_sig=="Enhancer_Skew")
lead.snp.dat <- data.frame(table(emvars_only$lead_snp))
names(lead.snp.dat) <- c("lead_snp", "number_of_emVars")
lead.snp.dat<-lead.snp.dat[order(-lead.snp.dat$number_of_emVars),]
# write.table(lead.snp.dat, paste0(data.dir,"/",mpra.name1,"_how_many_emvars_per_loci.txt"), row.names = F, col.names = T, sep="\t", quote=F) 

# What was the maximum number of emvars linked to a single lead variant and what was the lead variant?
lead.snp.dat.all <- data.frame(table(mpra$lead_snp))
names(lead.snp.dat.all) <- c("lead_snp", "number_of_variants")
lead.snp.dat.all<-lead.snp.dat.all[order(-lead.snp.dat.all$number_of_variants),]
write.table(lead.snp.dat.all, paste0(data.dir,"/",mpra.name1,"_how_many_variants_per_loci_all_types.txt"), row.names = F, col.names = T, sep="\t", quote=F) 
```

# PICS analysis
```{r pics enrichment}
# Change name
mpra <- mpra_merge
# Do it again
  mpra.pics.plot <- mpra
# order of gwas diseases
  gwas.order<- c("Crohns","MS","Psoriasis", "RA","T1D","UC", "IBD" )

# Format the mpra.pics.plot data
  # replace _CS_ with _PP_
  names(mpra.pics.plot)<-gsub("_CS_", "_PP_", names(mpra.pics.plot))
  # Select only certain columns
  mpra.pics.plot<-subset(mpra.pics.plot, project=="TGWAS", select=c(chr, snp_end, ld_snp, lead_snp, r2,                                       rsid,Crohns_pval,Crohns_pics,Crohns_PP_running,MS_pval,MS_pics,MS_PP_running,
                            Psoriasis_pval,Psoriasis_pics,Psoriasis_PP_running,RA_pval,RA_pics,RA_PP_running,
                            T1D_pval,T1D_pics,T1D_PP_running,UC_pval,UC_pics,UC_PP_running,IBD_pval,IBD_pics,
                            IBD_PP_running, dhs_Tcell_merged, dhs_all, mpra_sig))
  mpra.pics.plot$dhs_merged <- mpra.pics.plot$dhs_Tcell_merged

  
  # Remove bad SNPs where it doesn't reach 5E-8 association p-value in the GWAS and remove MHC region. These are hg19 SNPs # Added the loci with 3000+ variants
  bad_snps<-c("22:50966914:T:C","3:105558837:G:A", "12:9905851:A:C",
            "13:40745693:G:A","16:1073552:A:G","17:38775150:C:T",
            "17:44073889:A:G","18:12830538:G:A","2:100764087:T:G",
            "21:36488822:T:C","21:45621817:A:G","6:127457260:A:G",
            "6:130348257:C:T","7:116895163:G:A","7:51028987:T:A",
            "2:204592021:G:A", "14:75961511:C:T")
  mpra.pics.plot<-subset(mpra.pics.plot,  !(chr=="chr6" & snp_end>29691116 & snp_end<33054976) & !(lead_snp%in%bad_snps))
  
# For each mpra variant, find the disease with the strongest association and its associated PICS data
  mpra.pics.plot$top_pval<-NA #Top GWAS p-value for the MPRA variant
  mpra.pics.plot$top_disease<-NA #Disease corresponding to top GWAS p-value
  mpra.pics.plot$top_PP_running<-NA #Cummulative sum of posterior probabilities for that variant
  mpra.pics.plot$top_pics<-NA #PICS probability for that variant in the top GWAS

  for(i in 1:nrow(mpra.pics.plot)){ #Run through each MPRA variant
  
  top_pval<-max(mpra.pics.plot[i,grepl("_pval",names(mpra.pics.plot))], na.rm=T) #Find the top GWAS p-value
  top_disease<-str_split_fixed(names(mpra.pics.plot)[which(mpra.pics.plot[i,]==top_pval)][1], "\\_", 2)[1] #Find the disease corresponding to the top GWAS p-value
  
  #Write out GWAS and PICS data for top GWAS p-value
  mpra.pics.plot[i,]$top_pval<-top_pval
  mpra.pics.plot[i,]$top_disease<-top_disease
  mpra.pics.plot[i,]$top_PP_running<-mpra.pics.plot[i,paste0(top_disease, "_PP_running")]
  mpra.pics.plot[i,]$top_pics<-mpra.pics.plot[i,paste0(top_disease, "_pics")]
}
  mpra.pics.plot$top_pics<-as.numeric(mpra.pics.plot$top_pics)
  mpra.pics.plot$top_PP_running<-as.numeric(mpra.pics.plot$top_PP_running)

  dat.pics<-mpra.pics.plot

# Calculate MPRA/DHS enrichment in PICS fine-mapping.

  # creating an empty data frame
  dat.enrichment<-data.frame(pics=rep(c(0.01, 0.05, 0.1,0.15,0.2, 0.25,0.3,0.4,0.5,0.6,0.7,0.8), times=3), 
                     disease=rep(rep(c("all"), times=9), times=4), 
                     mpra=rep(c("mpra", "dhs", "mpra_dhs"), each=12), 
                     a=0, b=0, c=0, d=0,fold=0, p=0,odds=0,lower.conf=0,upper.conf=0, stringsAsFactors = F)

  # for loop for the stats
  for(i in 1:nrow(dat.enrichment)){
  if(dat.enrichment[i,]$mpra=="mpra"){ #Calculate MPRA enrichments in PICS fine-mapping
    a<-nrow(subset(dat.pics, mpra_sig=="Enhancer_Skew" &  top_pics > dat.enrichment[i,]$pics)) #emVar SNP with PICS fine-mapped
    b<-nrow(subset(dat.pics, mpra_sig=="Enhancer_Skew" &  top_pics <= dat.enrichment[i,]$pics)) #emVar SNP, but PICS not fine-mapped
    c<-nrow(subset(dat.pics, mpra_sig!="Enhancer_Skew" &  top_pics > dat.enrichment[i,]$pics)) #Not emVar SNP, but PICS fine-mapped
    d<-nrow(subset(dat.pics, mpra_sig!="Enhancer_Skew" &  top_pics <= dat.enrichment[i,]$pics)) #Not emVar SNP, and not PICS fine-mapped
  }
  if(dat.enrichment[i,]$mpra=="dhs"){ #Calculate DHS enrichments in PICS fine-mapping
    a<-nrow(subset(dat.pics, dhs_merged==1 &   top_pics > dat.enrichment[i,]$pics )) #DHS peak overlapping PICS fine-mapped SNP
    b<-nrow(subset(dat.pics, dhs_merged==1  &  top_pics <= dat.enrichment[i,]$pics)) #DHS peak not overlapping PICS fine-mapped SNP
    c<-nrow(subset(dat.pics, dhs_merged==0 &  top_pics > dat.enrichment[i,]$pics)) #Not overlapping DHS peak, but PICS fine-mapped SNP
    d<-nrow(subset(dat.pics, dhs_merged==0 &  top_pics <= dat.enrichment[i,]$pics)) #Not overlapping DHS peak and not PICS fine-mapped
  }
  if(dat.enrichment[i,]$mpra=="mpra_dhs"){ #Calcualte MPRA+DHS enrichments in PICS fine-mapping
    a<-nrow(subset(dat.pics, mpra_sig=="Enhancer_Skew" & dhs_merged==1 &   top_pics > dat.enrichment[i,]$pics )) #emVar, overlapping DHS peak and PICS fine-mapped
    b<-nrow(subset(dat.pics, mpra_sig=="Enhancer_Skew" & dhs_merged==1  &  top_pics <= dat.enrichment[i,]$pics)) #emVar and overlapping DHS peak, but not PICS fine-mapped
    c<-nrow(subset(dat.pics, (mpra_sig!="Enhancer_Skew" | dhs_merged==0) &  top_pics> dat.enrichment[i,]$pics)) #Either not emVar or not overlapping DHS peak, but PICS fine-mapped
    d<-nrow(subset(dat.pics, (mpra_sig!="Enhancer_Skew" | dhs_merged==0) &  top_pics <= dat.enrichment[i,]$pics)) #Either not emVar or not overlapping DHS peak, and not PICS fine-mapped
  }
  
  # Write out data
  dat.enrichment[i,]$a<-a
  dat.enrichment[i,]$b<-b
  dat.enrichment[i,]$c<-c
  dat.enrichment[i,]$d<-d
  dat.enrichment[i,]$fold<-(a/(a+b))/(c/(c+d)) #Calculate fold enrichment
  dat.enrichment[i,]$p<-fisher.test(rbind(c(a,b), c(c, d)))$p.value #Calculate enrichment p-value
  dat.enrichment[i,]$odds<-fisher.test(rbind(c(a,b), c(c, d)))$estimate #Calculate odds ratio
  dat.enrichment[i,]$lower.conf<-fisher.test(rbind(c(a,b), c(c, d)))$conf.int[1] #Calculate lower confidence interval
  dat.enrichment[i,]$upper.conf<-fisher.test(rbind(c(a,b), c(c, d)))$conf.int[2] #Calculate higher confidence interval
}
  
  
# This code plots the fold enrichment values against the PICS threshold (which are the p-values (the higher the threshold value, the lower the p-value)).
  dat.enrichment$pics<-factor(dat.enrichment$pics, levels=c(0.01, 0.05, 0.1,0.15,0.2, 0.25,0.3,0.4,0.5,0.6,0.7,0.8))
  dat.enrichment<-subset(dat.enrichment, pics%in%c(0.01,0.05, 0.1,0.2, 0.3,0.4,0.5,0.6,0.7,0.8 ))
# Write the dat in a table
write.table(dat.enrichment, paste0(data.dir,"/",mpra.name1,"_mpra_pics_enrichment_plot_all_loci_table.txt"), row.names = F, col.names = T, sep="\t", quote=F)  
# Create a plot of  emVar and DHS enrichment for PICS probable causal variants
   PICS_threshold_plot <-  ggplot(dat.enrichment, aes( y=fold, x=pics)) + 
    geom_bar(position="dodge", stat="identity", color="black",aes(fill=-log10(p))) +  scale_fill_gradient(low= "lightgray",high="#E14A4A",limits = c(0, max(-log10(dat.enrichment$p)))) +
    facet_grid(~mpra)+
    theme_bw()+ 
    geom_hline(yintercept=1, linetype="dashed", color = "black")+
    xlab("PICS threshold")+
    ylab("Fold Enrichment")+
    geom_text(aes(label=a), position="dodge", vjust=1)+
    geom_text(aes(label=round(-log10(p),2)), position="dodge", vjust=-0.75) +
    ggtitle("emVar enrichment and DHS enrichment for PICS probable causal variants", subtitle =mpra1_name) 
  PICS_threshold_plot
  
#########################

# emVars loci plot
  
#########################

# Subset the data to only loci with at least one emvar
emvar_lead_snps<-unique(subset(dat.pics, mpra_sig=="Enhancer_Skew")$lead_snp)
dat.pics<-subset(dat.pics, lead_snp%in%emvar_lead_snps)

  
 # Calculate MPRA/DHS enrichment in PICS fine-mapping.

  # creating an empty data frame
  dat.enrichment<-data.frame(pics=rep(c(0.01, 0.05, 0.1,0.15,0.2, 0.25,0.3,0.4,0.5,0.6,0.7,0.8), times=3), 
                     disease=rep(rep(c("all"), times=9), times=4), 
                     mpra=rep(c("mpra", "dhs", "mpra_dhs"), each=12), 
                     a=0, b=0, c=0, d=0,fold=0, p=0,odds=0,lower.conf=0,upper.conf=0, stringsAsFactors = F)

  # for loop for the stats
  for(i in 1:nrow(dat.enrichment)){
  if(dat.enrichment[i,]$mpra=="mpra"){ #Calculate MPRA enrichments in PICS fine-mapping
    a<-nrow(subset(dat.pics, mpra_sig=="Enhancer_Skew" &  top_pics > dat.enrichment[i,]$pics)) #emVar SNP with PICS fine-mapped
    b<-nrow(subset(dat.pics, mpra_sig=="Enhancer_Skew" &  top_pics <= dat.enrichment[i,]$pics)) #emVar SNP, but PICS not fine-mapped
    c<-nrow(subset(dat.pics, mpra_sig!="Enhancer_Skew" &  top_pics > dat.enrichment[i,]$pics)) #Not emVar SNP, but PICS fine-mapped
    d<-nrow(subset(dat.pics, mpra_sig!="Enhancer_Skew" &  top_pics <= dat.enrichment[i,]$pics)) #Not emVar SNP, and not PICS fine-mapped
  }
  if(dat.enrichment[i,]$mpra=="dhs"){ #Calculate DHS enrichments in PICS fine-mapping
    a<-nrow(subset(dat.pics, dhs_merged==1 &   top_pics > dat.enrichment[i,]$pics )) #DHS peak overlapping PICS fine-mapped SNP
    b<-nrow(subset(dat.pics, dhs_merged==1  &  top_pics <= dat.enrichment[i,]$pics)) #DHS peak not overlapping PICS fine-mapped SNP
    c<-nrow(subset(dat.pics, dhs_merged==0 &  top_pics > dat.enrichment[i,]$pics)) #Not overlapping DHS peak, but PICS fine-mapped SNP
    d<-nrow(subset(dat.pics, dhs_merged==0 &  top_pics <= dat.enrichment[i,]$pics)) #Not overlapping DHS peak and not PICS fine-mapped
  }
  if(dat.enrichment[i,]$mpra=="mpra_dhs"){ #Calcualte MPRA+DHS enrichments in PICS fine-mapping
    a<-nrow(subset(dat.pics, mpra_sig=="Enhancer_Skew" & dhs_merged==1 &   top_pics > dat.enrichment[i,]$pics )) #emVar, overlapping DHS peak and PICS fine-mapped
    b<-nrow(subset(dat.pics, mpra_sig=="Enhancer_Skew" & dhs_merged==1  &  top_pics <= dat.enrichment[i,]$pics)) #emVar and overlapping DHS peak, but not PICS fine-mapped
    c<-nrow(subset(dat.pics, (mpra_sig!="Enhancer_Skew" | dhs_merged==0) &  top_pics> dat.enrichment[i,]$pics)) #Either not emVar or not overlapping DHS peak, but PICS fine-mapped
    d<-nrow(subset(dat.pics, (mpra_sig!="Enhancer_Skew" | dhs_merged==0) &  top_pics <= dat.enrichment[i,]$pics)) #Either not emVar or not overlapping DHS peak, and not PICS fine-mapped
  }
  
  # Write out data
  dat.enrichment[i,]$a<-a
  dat.enrichment[i,]$b<-b
  dat.enrichment[i,]$c<-c
  dat.enrichment[i,]$d<-d
  dat.enrichment[i,]$fold<-(a/(a+b))/(c/(c+d)) #Calculate fold enrichment
  dat.enrichment[i,]$p<-fisher.test(rbind(c(a,b), c(c, d)))$p.value #Calculate enrichment p-value
  dat.enrichment[i,]$odds<-fisher.test(rbind(c(a,b), c(c, d)))$estimate #Calculate odds ratio
  dat.enrichment[i,]$lower.conf<-fisher.test(rbind(c(a,b), c(c, d)))$conf.int[1] #Calculate lower confidence interval
  dat.enrichment[i,]$upper.conf<-fisher.test(rbind(c(a,b), c(c, d)))$conf.int[2] #Calculate higher confidence interval
}
  
  
# This code plots the fold enrichment values against the PICS threshold (which are the p-values (the higher the threshold value, the lower the p-value)).
  dat.enrichment$pics<-factor(dat.enrichment$pics, levels=c(0.01, 0.05, 0.1,0.15,0.2, 0.25,0.3,0.4,0.5,0.6,0.7,0.8))
  dat.enrichment<-subset(dat.enrichment, pics%in%c(0.01,0.05, 0.1,0.2, 0.3,0.4,0.5,0.6,0.7,0.8 ))
# Write data into a table  
write.table(dat.enrichment, paste0(data.dir,"/",mpra.name1,"_mpra_pics_enrichment_plot_dhs_loci_only_table.txt"), row.names = F, col.names = T, sep="\t", quote=F)
  
# Create a plot with emVar and DHS enrichment for PICS probable causal variants for only loci with at least one emvar 
  PICS_threshold_emvars_loci_plot <-  ggplot(dat.enrichment, aes( y=fold, x=pics)) + 
    geom_bar(position="dodge", stat="identity", color="black",aes(fill=-log10(p))) +  scale_fill_gradient(low= "lightgray",high="#E14A4A",limits = c(0, max(-log10(dat.enrichment$p)))) +
    facet_grid(~mpra) +
    theme_bw()+ 
    geom_hline(yintercept=1, linetype="dashed", color = "black")+
    xlab("PICS threshold")+
    ylab("Fold Enrichment")+
    geom_text(aes(label=a), position="dodge", vjust=1)+
    geom_text(aes(label=round(-log10(p),2)), position="dodge", vjust=-0.75) +
    ggtitle("emVar enrichment and DHS enrichment for PICS probable causal variants (emVars loci only plot)", subtitle = mpra1_name) 
  PICS_threshold_emvars_loci_plot
  
  
# creating the plot name from the mpra.name1
plots_name1 <- paste0(mpra.name1, "_","PICS_threshold",".pdf")
# Saving the plot in the plots directory 
ggsave(plots_name1,plot = PICS_threshold_plot, path = plots.dir, width = 10, height = 5) 
# Do the same for the second plot
plots_name2 <- paste0(mpra.name1, "_", "PICS_threshold_emvars_loci",".pdf")
ggsave(plots_name2,plot = PICS_threshold_emvars_loci_plot, path = plots.dir, width = 10, height = 5) 



### Sensitivity and specificity calculations ###
dat.pics<-mpra.pics.plot
dhs_loci<-F #TRUE if calculation only for loci where a GWAS SNP overlaps a DHS peak
if(dhs_loci==T){  
  dat.pics<-subset(dat.pics, lead_snp%in%subset(dat.pics, dhs_Tcell_merged>0)$lead_snp)
}

#Calculate credible sets
dat.pics<-dat.pics[order(dat.pics$lead_snp, -dat.pics$top_pics),] 
dat.pics<-subset(dat.pics, select=c(ld_snp, lead_snp, r2, top_PP_running, top_pics, dhs_all, dhs_Tcell_merged, mpra_sig))
dat.pics$CS_80<-0 
dat.pics$CS_90<-0 
dat.pics$CS_95<-0 
for(i in 1:nrow(dat.pics)){
  top_pics<-max(subset(dat.pics, lead_snp==dat.pics[i,]$lead_snp)$top_pics)

  if(dat.pics[i,]$top_pics==top_pics){
    dat.pics[i,]$CS_80<-1
    dat.pics[i,]$CS_90<-1 
    dat.pics[i,]$CS_95<-1 
  }else{
    if(dat.pics[i,]$top_pics>=0.01){
      if(dat.pics[i,]$top_PP_running<=0.8){
        dat.pics[i,]$CS_80<-1
      }    
      if(dat.pics[i,]$top_PP_running<=0.9){
        dat.pics[i,]$CS_90<-1 
      }
      if(dat.pics[i,]$top_PP_running<=0.95){
        dat.pics[i,]$CS_95<-1 
      }
    }
  }
}

pics_cs_table <- dat.pics

# make an empty data frame 
dat.sens_spec<-data.frame(cs=rep(c(80,90,95)), disease="all", mpra=rep(c("mpra"), each=3), 
                          a=0, b=0, c=0, d=0,sensitivity=0, specificity=0, stringsAsFactors = F)

for(i in 1:nrow(dat.sens_spec)){
  a<-0 # Instantiate variable for MPRA emVars that are in PICS credible set 
  b<-0 #variable for MPRA emVars that are not PICS credible set 
  c<-0 #variable for MPRA non-emVars that are in PICS credible set 
  d<-0 #variable for MPRA non-emVars that are not PICS credible set 
  a_list<-c()
  b_list<-c()
  c_list<-c()
  d_list<-c()
  for(l in unique(dat.pics$lead_snp)){ #Perform this calculation for each GWAS locus
    temp.dat.pics<-subset(dat.pics, lead_snp==l) #Subset SNPs for that locus
    total_list<-nrow(temp.dat.pics)
    cs_list<-nrow(temp.dat.pics[temp.dat.pics[,paste0("CS_", dat.sens_spec[i,]$cs)]==1,])
    non_cs_list<-nrow(temp.dat.pics[temp.dat.pics[,paste0("CS_", dat.sens_spec[i,]$cs)]==0,])
    max_pics<-max(temp.dat.pics$top_pics)
    if(max_pics>=0.05 & cs_list<=25 & total_list>=5 & total_list<100 & non_cs_list>0 ){
      dat.cs<-temp.dat.pics[temp.dat.pics[,paste0("CS_", dat.sens_spec[i,]$cs)]==1,]
      if(nrow(subset(dat.cs, mpra_sig=="Enhancer_Skew"))>0){
        a<-a+1
        a_list<-c(a_list, l)
      }else{
        c<-c+1
        c_list<-c(c_list, l)
      }
      
      dat.non_cs<-temp.dat.pics[temp.dat.pics[,paste0("CS_", dat.sens_spec[i,]$cs)]==0,]
      if(nrow(subset(dat.non_cs, mpra_sig=="Enhancer_Skew"))>0){
        b<-b+1
        b_list<-c(b_list, l)
      }else{
        d<-d+1
        d_list<-c(d_list, l)
      }
    }
  }
  dat.sens_spec[i,]$a<-a
  dat.sens_spec[i,]$b<-b
  dat.sens_spec[i,]$c<-c
  dat.sens_spec[i,]$d<-d
  dat.sens_spec[i,]$sensitivity<-a/(a+c) #Calculate sensitivity
  dat.sens_spec[i,]$specificity<-d/(b+d) #Calculate specificity
}

write.table(dat.sens_spec, paste0(data.dir,"/",mpra.name1,"_dhsloci_",dhs_loci,"_","sensitivity_and_specificity"), row.names = F, col.names = T, sep="\t", quote=F)

# Sensitivity and specificity of not precision and recall!
# https://towardsdatascience.com/should-i-look-at-precision-recall-or-specificity-sensitivity-3946158aace1

```


# DHS enrichment plot
```{r dhs enrichment}
dhs.pos<- read.delim(gzfile(paste0(data.dir, "hg19_dhs/DHS_Index_and_Vocabulary_hg19_WM20190703.txt.gz")), header=T, stringsAsFactors = F)

# sample.dat
sample.dat<-read.delim(paste0(data.dir, "hg19_dhs/DHS_Index_and_Vocabulary_metadata.tsv"), sep="\t", header=T, stringsAsFactors = F)

# bring in dhs data
dhs.dat<-fread(paste0(data.dir, "hg19_dhs/dat_bin_FDR01_hg19.txt.gz"))

mpra <- mpra_merge

# Format mpra data
  # subset data
  mpra<-subset(mpra, project=="TGWAS", select=c(chr, snp_start, snp_end, mpra_sig))
  # Create strand column
  mpra$strand<-"+"
  # Make into a Granges file
  mpra.se<-makeGRangesFromDataFrame(mpra,  seqnames = "chr", start.field = "snp_start", end.field = "snp_end",keep.extra.columns=TRUE)
  
# format dhs.pos data
  # subset data
  dhs.pos<-subset(dhs.pos, select=c(seqname, start, end)) 
  # Make a peak name for easier indexing
  dhs.pos$peak<-paste0(dhs.pos$seqname, ":", dhs.pos$start, "_", dhs.pos$end) 
  # make a granges object out of the dataframe
  dhs.pos.se<-makeGRangesFromDataFrame(dhs.pos, seqnames = "seqname", start.field = "start", end.field = "end",keep.extra.columns=TRUE) #Make into GRanges format
  
# Find DHS peaks that overlap with MPRA
  #Find DHS peaks that overlap with MPRA SNPs
  overlap_peaks<-data.frame(subsetByOverlaps(dhs.pos.se, mpra.se, type="any"), stringsAsFactors = F)$peak 
  #Extract row indices of overlapping peaks from DHS file
  overlap_index<-which(dhs.pos$peak%in%overlap_peaks) 
  #Subset DHS peaks file by index of overlapping peaks
  dhs.pos<-dhs.pos[overlap_index,] 
  # remove the previous objects
  rm(mpra.se)
  rm(dhs.pos.se)
  
  dhs.dat<-dhs.dat[overlap_index,] #Subset DHS matrix file by index of overlapping peaks
  dhs.dat<-cbind(dhs.pos[,c(1:3)],dhs.dat)
  
# Format DHS sample information
  sample.dat<-subset(sample.dat, !is.na(library.order),select=c(library.order,Biosample.name, System,Subsystem,  Organ, Biosample.type, Biological.state))

# Merge all these data sets
  # For each cell type, merge DHS peaks from that cell type. Add a column for each cell type, and list it's peak membership    for any sample in that cell type
  dhs.merge.dat<-dhs.dat[,c(1:3)]
  for(c in unique(sample.dat$Biosample.name)){
  libraries<-paste0("V", subset(sample.dat, Biosample.name==c)$library.order)
  if(length(libraries)>1){
    dhs.merge.dat$newcol<-apply(dhs.dat[,c(libraries)],1, max)
  }else{
    dhs.merge.dat$newcol<-dhs.dat[,c(libraries)]
  }
  names(dhs.merge.dat)<-gsub("newcol", c, names(dhs.merge.dat))
  }
  
# Generate peak indices for MPRA SNPs for faster downstream processing
  dhs.pos<-dhs.merge.dat[,c(1:3)]
  dhs.pos$index<-seq(1:nrow(dhs.pos))
  dhs.merge.dat$index<-seq(1:nrow(dhs.merge.dat))
  mpra$peak_index<-NA
    for(i in 1:nrow(mpra)){ #Add in DHS peak index for any peak overlapping MPRA SNP
    mpra[i,]$peak_index<-subset(dhs.pos, seqname==mpra[i,]$chr & start<=mpra[i,]$snp_end & end >=mpra[i,]$snp_end)$index[1]  
      }
  pcre_index<-subset(mpra, mpra_sig%in%c("Enhancer_nSkew", "Enhancer_Skew") & !is.na(peak_index))$peak_index #peak indices     for MPRA pCRE SNPs only
  non_pcre_index<-subset(mpra, mpra_sig=="nEnhancer_nSkew" & !is.na(peak_index))$peak_index #peak indices for MPRA non-pCRE    SNPs
  
# Fixing sample names
  sample.dat$Biosample.name<-gsub("786_O","X786_O", sample.dat$Biosample.name)
  sample.dat$Biosample.name<-gsub("SK\\-N\\-DZ\\+RA\\s\\(3 days\\)","SK.N.DZ.RA..3.days.", sample.dat$Biosample.name)
  sample.dat$Biosample.name<-gsub("Namalwa\\s\\+\\sSendai\\svirus_2h","Namalwa...Sendai.virus_2h", sample.dat$Biosample.name)
  sample.dat$Biosample.name<-gsub("MCF10a_ER_SRC_6h_Tam/Washout18h","MCF10a_ER_SRC_6h_Tam.Washout.18.h",sample.dat$Biosample.name)


# Calculate enrichments of MPRA pCREs in DHS peaks for each cell type
  dat.plot<-data.frame(Biosample.name= unique(sample.dat$Biosample.name), System=NA, Subsystem=NA, pcre_fold=0, pcre_p=0, a=0, b=0, c=0, d=0, odds=0, lower.conf=0, upper.conf=0, stringsAsFactors = F)
  
  for(i in 1:nrow(dat.plot)){
  
  #Extract system for that cell type
  dat.plot[i,]$System<-subset(sample.dat, Biosample.name==dat.plot[i,]$Biosample.name)$System[1]
  
  #Extract subsystem for that cell type
  if(dat.plot[i,]$System!="Hematopoietic"){
    dat.plot[i,]$Subsystem<-dat.plot[i,]$System
  }else{
    dat.plot[i,]$Subsystem<-paste0("Heme_", subset(sample.dat, Biosample.name==dat.plot[i,]$Biosample.name)$Subsystem[1])
  }
  if(dat.plot[i,]$Subsystem=="Heme_"){dat.plot[i,]$Subsystem<-"Heme_Other"}
  if(dat.plot[i,]$Subsystem=="Fetal Life Support"){dat.plot[i,]$Subsystem<-"Fetal"}
  
  #Find the column in the dhs data that correspond to that cell type
  cell_col<-which(names(dhs.merge.dat)==dat.plot[i,]$Biosample.name)
  
  #find which peak_indexes for the peaks in that cell type
  peak_index<-dhs.merge.dat[dhs.merge.dat[,cell_col]==1,]$index

  #Calculate enrichment
  a<-length(intersect(peak_index, pcre_index)) #Number of pCRE in DHS peak
  b<-length(intersect(peak_index, non_pcre_index)) #Number of non-pCRE in DHS peak
  c<-nrow(subset(mpra, mpra_sig%in%c("Enhancer_nSkew", "Enhancer_Skew")))-a #Number of pCRE not in DHS
  d<-nrow(subset(mpra, mpra_sig=="nEnhancer_nSkew"))-b #Number of non-pCRE not in DHS
  dat.plot[i,]$a<-a
  dat.plot[i,]$b<-b
  dat.plot[i,]$c<-c
  dat.plot[i,]$d<-d
  dat.plot[i,]$pcre_fold<-(a/(a+c))/(b/(b+d)) #Calcualte fold enrichment
  dat.plot[i,]$pcre_p<-fisher.test(rbind(c(a,b), c(c, d)), alternative="two.sided")$p.value #Calculate p-value
  dat.plot[i,]$odds<-fisher.test(rbind(c(a,b), c(c, d)))$estimate #Calculate odds ratio
  dat.plot[i,]$lower.conf<-fisher.test(rbind(c(a,b), c(c, d)))$conf.int[1] #Calculate lower confidence interval
  dat.plot[i,]$upper.conf<-fisher.test(rbind(c(a,b), c(c, d)))$conf.int[2] #Calculate higher confidence interval
  }
  
# Rank data by p-value
  dat.plot<-dat.plot[order(dat.plot$pcre_p),]
  dat.plot$rank<-seq(1:nrow(dat.plot))
  
# Plot of ranked p-values
  # list of biological systems to color code by
subsystem.order<-c("Cardiovascular","Connective","Digestive","Embryonic","Endocrine","Epithelial","Fetal","Genitourinary", "Hepatic","Integumentary","Musculoskeletal","Nervous","Renal","Respiratory","Heme_Erythroid","Heme_Myeloid","Heme_Other","Heme_Progenitor","Heme_Lymphoid","Heme_T-cell")
  # make subsystem a factor in the plot
dat.plot$Subsystem<-factor(dat.plot$Subsystem, levels=subsystem.order)
  
# ranking the plot data
dat.plot<-dat.plot[order(dat.plot$pcre_p),]
dat.plot$rank<-seq(1:nrow(dat.plot))

write.table(dat.plot, paste0(data.dir,mpra.name1,"_encode_dhs_enrichment_table.txt"), row.names=F, col.names=T, sep="\t", quote=F)


ylimit <- max(-log10(dat.plot$pcre_p)) + 1

p_value_rank_dhs_plot <- ggplot()+geom_point(data=dat.plot[!(grepl("Heme", dat.plot$Subsystem)),],aes(x=rank,y=-log10(pcre_p)),color="grey", size=1.5, pch=16)+
  geom_point(data=dat.plot[(grepl("Heme", dat.plot$Subsystem)),],aes(x=rank,y=-log10(pcre_p), color=Subsystem), size=3, pch=16)+coord_cartesian(ylim=c(0,ylimit), xlim=c(0,nrow(dat.plot)))+scale_color_manual(values=brewer.pal(6, "Dark2"))+theme_classic()+  geom_text_repel(data=subset(dat.plot, as.numeric(rank)<6),aes(x=rank,y=-log10(pcre_p),label=Biosample.name),nudge_x = 10, nudge_y = 0.5, hjust=0) + ggtitle("pCRE ranked p-values in DHS peaks", subtitle=mpra1_name)
p_value_rank_dhs_plot 

# creating the plot name from the mpra.name1
plots_name <- paste0(mpra.name1,"_p_value_rank_dhs",".pdf")
# Saving the plot in the plots directory 
ggsave(plots_name,plot = p_value_rank_dhs_plot, path = plots.dir, width = 7, height = 7)
``` 


# Allelic Skew 
```{r allelic skew}
mpra <- mpra_merge
# This is 
dat.skew <- data.frame(read_excel(paste0(data.dir,"41588_2019_505_MOESM3_ESM.xlsx"), sheet=10), stringsAsFactors = F)

dat.skew<-dat.skew[!grepl(paste0("\\-",eliminate_stim_status), dat.skew$cell_type),] # if analyzing stimulated cells remove unstimulated cell types

# Parse ATAC allelic skew file
dat.skew$chr<-str_split_fixed(dat.skew$het_id, "\\_", 5)[,2]
dat.skew$pos<-as.numeric(str_split_fixed(dat.skew$het_id, "\\_", 5)[,3])
dat.skew$chr_pos<-paste(dat.skew$chr, dat.skew$pos, sep=":")

#Collapse allelic skew counts across cell types for each SNP
dat.skew.collapse<-data.frame(chr_pos=unique(dat.skew$chr_pos), het_id=0,atac_ref=0, atac_alt=0, stringsAsFactors = F)
for(i in 1:nrow(dat.skew.collapse)){
  dat.skew.collapse[i,]$atac_ref<-sum(subset(dat.skew, chr_pos==dat.skew.collapse[i,]$chr_pos)$refCount)
  dat.skew.collapse[i,]$atac_alt<-sum(subset(dat.skew, chr_pos==dat.skew.collapse[i,]$chr_pos)$altCount)
  dat.skew.collapse[i,]$het_id<-subset(dat.skew, chr_pos==dat.skew.collapse[i,]$chr_pos)$het_id[1]
}

# Subsetting the mpra data
#mpra<-subset(mpra, project=="TGWAS" & !is.na(LogSkew),select=c(chr, snp_start, snp_end, LogSkew, mpra_sig))

#Merge ATAC allelic skew data with MPRA data
mpra$chr_pos<-paste0(mpra$chr, ":", mpra$snp_end)
mpra<-merge(mpra, dat.skew.collapse, by="chr_pos", all.x=T, all.y=F)
mpra$allelic_skew<-mpra$atac_alt/(mpra$atac_alt+mpra$atac_ref)

#Plot MPRA Skew against ATAC allelic skew
mpra.skew<-subset(mpra, mpra_sig%in%c("Enhancer_Skew" ,"Enhancer_nSkew") & !is.na(het_id) )

#Generate correlation statistics
skew_vs_skew_model <- summary(lm(mpra.skew$LogSkew~mpra.skew$allelic_skew))
skew_vs_skew_model 

# Make the correlation plot
MPRA_skew_vs_allelic_skew_plot<- plot(x=mpra.skew$allelic_skew, y=mpra.skew$LogSkew, pch=16, cex=1.2 ,col="darkgray", xlim=c(0,1.2), ylim=c(-1,1.2))
points(x=mpra.skew[mpra.skew$mpra_sig=="Enhancer_Skew",]$allelic_skew, y=mpra.skew[mpra.skew$mpra_sig=="Enhancer_Skew",]$LogSkew, pch=16, cex=1.2 ,col="red", xlim=c(0,1), ylim=c(-1,1))
abline(v=0.5, lty=2, col="blue")
abline(h=0, lty=2, col="blue")
abline(lm(mpra.skew$LogSkew~mpra.skew$allelic_skew))
print(MPRA_skew_vs_allelic_skew_plot)
MPRA_skew_vs_allelic_skew_ggplot <- ggplot(data=mpra.skew, aes(x=allelic_skew, y=LogSkew)) + geom_point(color="darkgrey") + theme_bw()  + geom_hline(yintercept=0,color="#5D93CD") +  geom_vline(xintercept=0.5,color="#5D93CD") + geom_point(data=subset(mpra.skew, mpra_sig=="Enhancer_Skew"),color="#E14A4A") + geom_smooth(method = "lm",color="#5D93CD",fill="#5D93CD") + labs(title="mpra skew vs. allelic atac skew correlation plot", subtitle=mpra1_name, caption= paste0("r^2: ", skew_vs_skew_model$r.squared, "; p-value: ", skew_vs_skew_model$coefficients[2,4])) + xlim(0,1) + ylim(-2.25,2.25)

print(MPRA_skew_vs_allelic_skew_ggplot)

# creating the plot name from the mpra.name1
plots_name <- paste0(mpra.name1,"_LogSkew_Allelic_Skew_correlation",".pdf")
# Saving the plot in the plots directory 
ggsave(plots_name,plot = MPRA_skew_vs_allelic_skew_ggplot, path = plots.dir, width = 7, height = 7)

#Set up the data for the plot of proportion of each variant category with allelic skew
# Create an empty dataframe
dat.plot<-data.frame( mpra_sig=c("nEnhancer_nSkew", "Enhancer_nSkew", "Enhancer_Skew"), proportion=0,total=0,stringsAsFactors = F)
# Calculate the proportions
for(i in 1:nrow(dat.plot)){
  dat.plot[i,]$proportion<-nrow(subset(mpra, !is.na(allelic_skew) & mpra_sig==dat.plot[i,]$mpra_sig))/nrow(subset(mpra, mpra_sig==dat.plot[i,]$mpra_sig))
}
# count the totals
for(i in 1:nrow(dat.plot)){
  dat.plot[i,]$total<-nrow(subset(mpra, !is.na(allelic_skew) & mpra_sig==dat.plot[i,]$mpra_sig))
}
# Make variant category a factor
dat.plot$mpra_sig<-factor(dat.plot$mpra_sig, levels=c("nEnhancer_nSkew", "Enhancer_nSkew", "Enhancer_Skew"))


ylimit <- max(dat.plot$proportion) + 0.01

#Calculate enrichment p-values of MPRA variants in allelic skew variants
#For pCREs
proportion_test_pcre <- prop.test(x = c(nrow(subset(mpra, !is.na(allelic_skew) & mpra_sig=="Enhancer_nSkew")), nrow(subset(mpra, !is.na(allelic_skew) & mpra_sig=="nEnhancer_nSkew"))), 
          n =c(nrow(subset(mpra,  mpra_sig=="Enhancer_nSkew")), nrow(subset(mpra, mpra_sig=="nEnhancer_nSkew"))))
proportion_test_pcre$p.value
proportion_test_pcre
#For emVars
proportion_test_emvar <- prop.test(x = c(nrow(subset(mpra, !is.na(allelic_skew) & mpra_sig=="Enhancer_Skew")), nrow(subset(mpra, !is.na(allelic_skew) & mpra_sig=="nEnhancer_nSkew"))), 
          n =c(nrow(subset(mpra,  mpra_sig=="Enhancer_Skew")), nrow(subset(mpra, mpra_sig=="nEnhancer_nSkew"))))
proportion_test_emvar$p.value
proportion_test_emvar


dat.plot$p <- c(NA, proportion_test_pcre$p.value, proportion_test_emvar$p.value)

#Plot proportion of pCREs and emVars that show allelic skew in Calderon ATAC-seq data
emVars_Skew_plot <- ggplot(dat.plot, aes(fill=mpra_sig, y=proportion, x=mpra_sig)) + 
  geom_bar(position="dodge", stat="identity", color="black")+
  theme_bw()+scale_fill_brewer(palette = "Blues")+coord_cartesian(ylim=c(0,ylimit))+
  theme(axis.text.x = element_text( angle = 45, hjust=1))+ 
  geom_hline(yintercept=1, linetype="dashed", color = "black") + ggtitle("Proportion of pCREs and emVars that show allelic skew in Calderon ATAC-seq data", subtitle = mpra1_name) + xlab("Variant category") +
  geom_text(aes(label=signif(p,2)), position="dodge", vjust=-0.75) +
   geom_text(aes(label=round(total,3)), position="dodge", vjust=1.5) 
 print(emVars_Skew_plot)

  # creating the plot name from the mpra.name1
plots_name <- paste0(mpra.name1,"_emVars_Allelic_Skew",".pdf")
# Saving the plot in the plots directory 
ggsave(plots_name,plot = emVars_Skew_plot, path = plots.dir, width = 7, height = 7)
```

# TSS plot
```{r tss}
mpra <- mpra_merge

txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene

# This subsets the MPRA data
  mpra<-subset(mpra, project=="TGWAS" & !is.na(LogSkew))
  
#Expand out MPRA data
  mpra.all<-mpra
  mpra.all$mpra_sig_expanded<-"all"
  mpra.pcre<-subset(mpra, mpra_sig%in%c("Enhancer_nSkew","Enhancer_Skew"))
  mpra.pcre$mpra_sig_expanded<-"pcre"
  mpra.emvar<-subset(mpra, mpra_sig=="Enhancer_Skew")
  mpra.emvar$mpra_sig_expanded<-"emvar"
  mpra.expanded<-rbind(mpra.all, mpra.pcre, mpra.emvar)

# This code uses the human reference genome we imported earlier to annotate our MPRA data with information about the function of different sections of the genome. Since it scans the big reference genome, it takes some time to run. annoDb="org.Hs.eg.db" is changed to NUll because it does not work.
#Annotate MPRA functional annotations
mpra.all.se<-makeGRangesFromDataFrame(mpra.expanded,  seqnames = "chr", start.field = "snp_start", end.field = "snp_end",keep.extra.columns=TRUE)
# peakAnno <- as.data.frame(annotatePeak(mpra.all.se, tssRegion=c(-3000, 3000), TxDb=txdb, annoDb="org.Hs.eg.db"))
peakAnno <- as.data.frame(annotatePeak(mpra.all.se, tssRegion=c(-3000, 3000), TxDb=txdb, annoDb=NULL))

#Creating the data for the upcoming plot
dat.plot<-data.frame(start=c(0, 1000, 3000, 10000, 20000 ),
                     end=c(1000, 3000, 10000,  20000,1000000), 
                     all=0, pcre=0, emvar=0, stringsAsFactors = F)

for(i in 1:nrow(dat.plot)){
  dat.plot[i,]$all<-nrow(subset(peakAnno, distanceToTSS>dat.plot[i,]$start & distanceToTSS<=dat.plot[i,]$end ))/nrow(peakAnno)
  dat.plot[i,]$pcre<-nrow(subset(peakAnno, mpra_sig%in%c("Enhancer_nSkew", "Enhancer_Skew") & distanceToTSS>dat.plot[i,]$start & distanceToTSS<=dat.plot[i,]$end ))/nrow(subset(peakAnno, mpra_sig%in%c("Enhancer_nSkew", "Enhancer_Skew")))
  dat.plot[i,]$emvar<-nrow(subset(peakAnno, mpra_sig%in%c("Enhancer_Skew") & distanceToTSS>dat.plot[i,]$start & distanceToTSS<=dat.plot[i,]$end ))/nrow(subset(peakAnno, mpra_sig%in%c("Enhancer_Skew")))
}

dat.plot$label=paste(dat.plot$start, dat.plot$end, sep="_")
dat.plot<-subset(dat.plot, select=-c(start, end))
dat.plot<-melt(dat.plot, id.vars="label")
dat.plot$label<-factor(dat.plot$label, levels=dat.plot$label[1:5])


TTSvsVariants_plot<- ggplot(dat.plot, aes(x=label, y=value, fill=variable))+ 
  geom_bar(stat="identity", width=0.9, position = "dodge") +
  theme_bw()+scale_fill_brewer(palette = "Blues")+
  theme(axis.text.x = element_text( angle = 45, hjust=1))+
  xlab("Distance to TSS (bp)") + ylab("Proportion of variants") + ggtitle("Distance from the MPRA variants to the transcription start sites", subtitle=mpra1_name) # +  geom_text(aes(label=round(value,2), group=variable),position = position_dodge(width=1), vjust=-.75)
TTSvsVariants_plot
  # creating the plot name from the mpra.name1
plots_name <- paste0(mpra.name1,"_TSSvsVariants",".pdf")
# Saving the plot in the plots directory 
ggsave(plots_name,plot = TTSvsVariants_plot, path = plots.dir, width = 7, height = 7)
```

# MPRA caQTL
```{r caqtl}
mpra <- mpra_merge

# dat.qtl (From Gate et al ATAC-QTL data (PMID: 29988122), Supplemental Table 6)
dat.qtl <- data.frame(read_excel(paste0(data.dir,"41588_2018_156_MOESM8_ESM.xlsx"), sheet=2), stringsAsFactors = F)

  mpra<-subset(mpra,project=="TGWAS", select=c(chr, snp_start, snp_end, LogSkew, mpra_sig))
  
# Format dat.qtl file
  # Creating chromosome column
  dat.qtl$chr<-str_split_fixed(dat.qtl$SNP, "\\:", 2)[,1]
  # creating end_snp column
  dat.qtl$snp_end<-str_split_fixed(str_split_fixed(dat.qtl$SNP, "\\_", 2)[,1], "\\:", 2)[,2]
  # subset dat.qtl data
  dat.qtl<-subset(dat.qtl, select=c(chr, snp_end, beta, p.value))
  # change the column names of dat.qtl data
  names(dat.qtl)[3:4]<-c("atac_qtl_beta", "atac_qtl_pval")
  
# Merge caQTL data with MPRA data
  mpra<-merge(mpra, dat.qtl, by=c("chr", "snp_end"), all.x=T, all.y=F)
  
# Plot data set-up
  # Calculates the proportion of variants in caqtls
  dat.plot<-data.frame(table(subset(mpra, !is.na(atac_qtl_pval))$mpra_sig)/table(mpra$mpra_sig),stringsAsFactors = F)
  names(dat.plot)<-c("mpra_sig", "proportion")
  # calculate total variants in caqtls
  totals<-data.frame(table(subset(mpra, !is.na(atac_qtl_pval))$mpra_sig),stringsAsFactors = F)
  dat.plot$totals <- totals$Freq
  # making variant category factor in the plots
  dat.plot$mpra_sig<-factor(dat.plot$mpra_sig, levels=c("nEnhancer_nSkew", "Enhancer_nSkew", "Enhancer_Skew"))
  
#Calculate enrichment p-values of MPRA variants in allelic skew variants
  #For pCREs
  caqtl_proportion_test_pcre <-  prop.test(x = c(nrow(subset(mpra, !is.na(atac_qtl_pval) & mpra_sig=="Enhancer_nSkew")), nrow(subset(mpra, !is.na(atac_qtl_pval)& mpra_sig=="nEnhancer_nSkew"))),n =c(nrow(subset(mpra,  mpra_sig=="Enhancer_nSkew")), nrow(subset(mpra, mpra_sig=="nEnhancer_nSkew"))))
caqtl_proportion_test_pcre  
  #For emVars
  caqtl_proportion_test_emvar <-  prop.test(x = c(nrow(subset(mpra, !is.na(atac_qtl_pval) & mpra_sig=="Enhancer_Skew")), nrow(subset(mpra, !is.na(atac_qtl_pval)& mpra_sig=="nEnhancer_nSkew"))),n =c(nrow(subset(mpra,  mpra_sig=="Enhancer_Skew")), nrow(subset(mpra, mpra_sig=="nEnhancer_nSkew"))))  
caqtl_proportion_test_emvar
dat.plot$p <- c(caqtl_proportion_test_pcre$p.value,caqtl_proportion_test_emvar$p.value,NA)
  
ylimit <- max(dat.plot$proportion) + 0.005
  
caQTL_emVars_Skew_plot<- ggplot(dat.plot, aes(fill=mpra_sig, x=mpra_sig, y=proportion)) + 
  geom_bar(position="dodge", stat="identity", color="black")+
  theme_bw()+scale_fill_brewer(palette = "Blues")+coord_cartesian(ylim=c(0,ylimit))+
  theme(axis.text.x = element_text( angle = 45, hjust=1))+ 
  geom_hline(yintercept=1, linetype="dashed", color = "black") + ggtitle("Proportion of MPRA variants that are caQTLs",     subtitle=mpra1_name) +
  geom_text(aes(label=signif(p,2)), position="dodge", vjust=-0.75)  +
  geom_text(aes(label=round(totals,3)), position="dodge", vjust=1.5) 
caQTL_emVars_Skew_plot


# Plot MPRA Skew against caQTL skew 
  # subset to only pcres and emvars
  mpra.qtl<-subset(mpra, !is.na(atac_qtl_pval) & mpra_sig%in%c("Enhancer_Skew", "Enhancer_nSkew"))
  
  # Linear model of mpra skew vs. caqtl skew
mpra_skew_caqtl_skew_model <- summary(lm(mpra.qtl$LogSkew~mpra.qtl$atac_qtl_beta))
mpra_skew_caqtl_skew_model

  # creating the plot
  mpra_skew_caqtl_skew_plot <- plot(x=mpra.qtl$atac_qtl_beta, y=mpra.qtl$LogSkew, pch=16, cex=1.2 ,col="darkgray", xlim=c(0.3,0.7), ylim=c(-1.5,1.5))
points(x=mpra.qtl[mpra.qtl$mpra_sig=="Enhancer_Skew",]$atac_qtl_beta, y=mpra.qtl[mpra.qtl$mpra_sig=="Enhancer_Skew",]$LogSkew, pch=16, cex=1.2 ,col="red", xlim=c(0,1), ylim=c(-1,1))
abline(v=0.5, lty=2, col="blue")
abline(h=0, lty=2, col="blue")
abline(lm(mpra.qtl$LogSkew~mpra.qtl$atac_qtl_beta))

mpra_skew_caqtl_skew_plot

# Create the correlation plot between MPRA and caQTL skew
mpra_skew_caqtl_skew_ggplot <- ggplot(data=mpra.qtl, aes(x=atac_qtl_beta, y=LogSkew)) + geom_point(color="darkgrey") + theme_bw()  + geom_hline(yintercept=0,color="#5D93CD") +  geom_vline(xintercept=0.5,color="#5D93CD") + geom_point(data=subset(mpra.qtl, mpra_sig=="Enhancer_Skew"),color="#E14A4A") + geom_smooth(method = "lm",color="#5D93CD",fill="#5D93CD") + labs(title="mpra skew vs. caqtl skew correlation plot", subtitle=mpra1_name, caption= paste0("r^2: ", mpra_skew_caqtl_skew_model$r.squared, "; p-value: ", mpra_skew_caqtl_skew_model$coefficients[2,4])) + xlim(0.25,0.75) + ylim(-1,1)
print(mpra_skew_caqtl_skew_ggplot)

# create the plot name from the mpra.name1
  plots_name <- paste0(mpra.name1,"_caQTL_Skew_emVars_LogSkew_correlation",".pdf")
# Saving the plot in the plots directory 
  ggsave(plots_name,plot = mpra_skew_caqtl_skew_ggplot, path = plots.dir, width = 7, height = 7)

# create the plot name from the mpra.name1
  plots_name <- paste0(mpra.name1,"_caQTL_emVars_Skew",".pdf")
# Saving the plot in the plots directory 
  ggsave(plots_name,plot = caQTL_emVars_Skew_plot, path = plots.dir, width = 7, height = 7)
```


# Delta SVM
```{r deltasvm}
mpra <- mpra_merge

if(mpra1_name=="Primary Tcell"){
  dat.deltasvm <- read.delim(paste0(mhguo.dir,"delta_svm/mpra_snps_E2_naive_CD4_deltaSVM.txt"), header=F, stringsAsFactors = F, sep="\t")
}

# format the input files
  # subset the mpra data
  mpra<-subset(mpra,project=="TGWAS" & mpra_sig=="Enhancer_Skew", select=c(chr, snp_start, snp_end, SNP, mpra_sig, LogSkew))
  # give different names to columns in deltasvm data
  names(dat.deltasvm)<-c("SNP", "delta_svm")
  # merge mpra and deltasvm data
  mpra.merge<-merge(mpra, dat.deltasvm, by="SNP", all.x=F, all.y=F)

# Linear model between delta svm score and logskew
logSkew_deltaSVM_model <- summary(lm(mpra.merge$delta_svm~mpra.merge$LogSkew))
logSkew_deltaSVM_model

# plot the correlation between MPRA  logSkew and deltaSVM score
  logSkew_deltaSVM_plot <- ggplot(mpra.merge, aes(x=delta_svm, y=LogSkew))+geom_point(color="red", fill="red",pch=16,  alpha=0.5, size=2)+
  theme_bw()+ 
  geom_hline(yintercept=0, linetype="dashed", color = "black")+
  geom_vline(xintercept=0, linetype="dashed", color = "black")+ 
  geom_abline(intercept = as.numeric(lm(mpra.merge$LogSkew~mpra.merge$delta_svm)$coefficients[1]), slope = as.numeric(lm(mpra.merge$LogSkew~mpra.merge$delta_svm)$coefficients[2]), color="blue")+
  coord_cartesian(xlim=c(-7.5,7.5), ylim=c(-2.5, 2.5)) + labs(title="logSkew vs. deltaSVM", subtitle=mpra1_name, caption=paste0("r^2: ",logSkew_deltaSVM_model$r.squared, "; p-value: ", logSkew_deltaSVM_model$coefficients[2,4]))
  logSkew_deltaSVM_plot
  
# creating the plot name from the mpra.name1
plots_name <- paste0(mpra.name1,"_logSkew_deltaSVM",".pdf")
# Saving the plot in the plots directory 
ggsave(plots_name,plot = logSkew_deltaSVM_plot, path = plots.dir, width = 7, height = 7)
```

# GWAS loci counts
```{r gwas loci}
mpra <- mpra_merge

# Format mpra data (call it mpra2 because you will need the oiginal mpra later)
  mpra2<-subset(mpra, project=="TGWAS",select=c( chr, snp_end,ld_snp, mpra_sig))
  
# Read in and merge LD files across chromosomes
mhguo_mpra_merge <- read.delim(paste0(mhguo.dir,"annotate_mpra/","mpra_data_merge.txt"), header=T, stringsAsFactors = F, sep="")
mhguo_mpra_merge <- mhguo_mpra_merge[!duplicated(mhguo_mpra_merge$SNP), ]
dat.ld.all <- subset(mhguo_mpra_merge, select=c(lead_snp,ld_snp,r2))
  
  # merge the mpra.chr data and the mpra data
  mpra.merge<-merge(dat.ld.all, mpra2, by="ld_snp", all.x=F, all.y=F)
  # Remove MHC region and also 17:44073889:A:G as they have very high number of MPRA variants tested.
  dat.sig<-subset(mpra.merge, !(chr==6 & snp_end>29691116 & snp_end<33054976) & !(lead_snp%in%c("17:44073889:A:G")) & r2>=0.8 & mpra_sig=="Enhancer_Skew" )
  
# plot histogram of mpra snps per locus
  # Create data for plot 
  dat.plot<-data.frame(table(dat.sig$lead_snp), stringsAsFactors = F)
  table1 <- table(dat.sig$lead_snp)
  # create plot with x limit of eight
  MPRA_SNP_per_Locus_plot <- hist(table1[table1 >= 0 & table1 < 8], xlim=c(0,8),breaks=seq(0,8,by=1), ylim=c(0,140), col="skyblue2", main="Histogram of # of MPRA SNPs per locus" ,xlab="# of MPRA SNPs", ylab="# of Loci")
MPRA_SNP_per_Locus_plot
  # create plot with x limit of 100
  MPRA_SNP_per_Locus_all <- hist(table1, xlim=c(0,100),breaks=seq(0,100,by=1), ylim=c(0,140), col="skyblue2", main="Histogram of # of MPRA SNPs per locus", xlab="# of MPRA SNPs", ylab="# of Loci")
MPRA_SNP_per_Locus_plot   
  
# Create ggplot version of this plot
ggplot_table <-  data.frame(table1)
ggplot_mpra_snp_per_locus_plot <- ggplot(data=ggplot_table, aes(x=Freq)) + geom_histogram(fill="#5D93CD") + theme_bw() + labs(title="Histogram of # of MPRA SNPs per locus",subtitle=mpra1_name,x="# of MPRA SNPs",y="# of Loci")
ggplot_mpra_snp_per_locus_plot

# What percent of loci have only 1 emVar (of loci with emVars)?
nrow(subset(ggplot_table, Freq==1))/nrow(ggplot_table)
# 0.5361702
  
# creating the plot name from the mpra.name1
plots_name <- paste0(mpra.name1,"_mpra_snps_per_locus_histogram",".pdf")
# Saving the plot in the plots directory 
ggsave(plots_name,plot = ggplot_mpra_snp_per_locus_plot, path = plots.dir, width = 7, height = 7)

# Plot of gwas loci explained by each disease 
  # bring in mpra again and subset
  mpra<-subset(mpra, project=="TGWAS")
  #Order of GWAS diseases
  gwas.order<- c("Crohns","MS","Psoriasis", "RA","T1D","UC", "IBD" ) 
  # Create dataframe with one row per disease. Then, count total number of loci tested, as well as # of loci with at least one emVar
  dat.disease<-data.frame(disease=gwas.order, total=0, mpra=0, stringsAsFactors = F)
  for(i in 1:nrow(dat.disease)){
  d<-dat.disease[i,]$disease
  temp<-mpra[!is.na(mpra[,which(names(mpra)==paste0(d, "_pval"))]) & mpra[,which(names(mpra)==paste0(d, "_pval"))]>=-log10(5e-8),]
  dat.disease[i,]$total<-length(unique(temp$lead_snp))
  dat.disease[i,]$mpra<-length(unique(temp[temp$mpra_sig=="Enhancer_Skew",]$lead_snp))
  }

  # Format data to use in the plot
  dat.disease<-dat.disease[order(-dat.disease$total),]
  dat.disease$total<-dat.disease$total-dat.disease$mpra
  dat.plot<-melt(dat.disease, id.vars="disease")
  dat.plot$disease<-factor(dat.plot$disease, levels=unique(dat.disease$disease))

proportion_loci_explained_plot <- ggplot(data=dat.plot[dat.plot$variable!="mpra_atac",], aes(x=disease, y=value,  fill =variable)) +geom_bar(stat="identity")+
  theme_bw()+scale_fill_brewer(palette = "Set2")+ylab("# of loci")+
  theme(axis.text.x=element_text(color = "black",angle=45,hjust=1)) + ggtitle("Proportion of loci explained by disease", subtitle=mpra1_name) # + geom_text(aes(label=round(value,3)), position="dodge", vjust=-.2) 
  # guides(fill=guide_legend(title="Loci", label.theme= element_text(c("Non-emVar","EmVar"))))
proportion_loci_explained_plot

# creating the plot name from the mpra.name1
plots_name <- paste0(mpra.name1,"_proportion_loci_explained",".pdf")
# Saving the plot in the plots directory 
ggsave(plots_name,plot = proportion_loci_explained_plot, path = plots.dir, width = 7, height = 7)

plots_name <- paste0(mpra.name1,"_proportion_loci_explained",".pdf")
# Saving the plot in the plots directory 
ggsave(plots_name,plot = proportion_loci_explained_plot, path = plots.dir, width = 7, height = 7)


```

# Histone Chromhmm
```{r histone}
mpra <- mpra_merge

# manifest.histone (manifest file containing file paths to ENCODE histone ChIP bed files for T cells)
manifest.histone<-data.frame(read_excel(paste0(mhguo.dir,"histone/histone_manifest.xlsx")), stringsAsFactors = F)

# cage
# I cannot download into R directly from the internet so I am going to download from the browser and upload to the server
# cage<-data.frame(fread("https://fantom.gsc.riken.jp/5/datafiles/latest/extra/Enhancers/human_permissive_enhancers_phase_1_and_2.bed.gz")) #read in CAGE data from FANTOM consortium
# Importing the data into R from the file which is now on the computer
cage<-data.frame(fread(paste0(data.dir, "human_permissive_enhancers_phase_1_and_2.bed.gz")))

# chromHMM file from ENCODE.
chromhmm<-fread("https://egg2.wustl.edu/roadmap/data/byFileType/chromhmmSegmentations/ChmmModels/core_K27ac/jointModel/final/E034_18_core_K27ac_dense.bed.gz", data.table = F, stringsAsFactors = F, header=F)

# Format mpra data
  mpra<-subset(mpra,project=="TGWAS", select=c(chr, snp_start, snp_end, ID, mpra_sig, dhs_Tcell_merged))
  mpra.se<-makeGRangesFromDataFrame(mpra, seqnames = "chr", start.field = "snp_start", end.field = "snp_end",keep.extra.columns=TRUE)

# Merging the manifest.histone data into mpra
  histone.list<-unique(manifest.histone$Histone)
  for(h in histone.list){ #Run for each histone mark
  bed_url.list<-subset(manifest.histone, Histone==h )$link #Grab urls for all bed files for that histone mark  
  bed<-data.frame(V1=character(), V2=numeric(), V3=numeric(), stringsAsFactors = F) #Create empty bed file
  for(url in bed_url.list){ #Read in and concatenate each bed file of the histone mark
    temp<-data.frame(fread(url), stringsAsFactors = F)
    temp<-temp[temp$V1%in%paste0("chr", seq(1,22)),c(1:3)]
    bed<-rbind(temp, bed)
  }
  names(bed)<-c("chr", "start", "end")
  #Create GRange for concatenated bed file
  bed.se<-makeGRangesFromDataFrame(bed, seqnames = "chr", start.field = "start", end.field = "end") 
  #Insert new column in MPRA file with whether MPRA variant overlaps the histone mark
  mpra$newcol<-ifelse(mpra$ID%in%(subsetByOverlaps( mpra.se, bed.se,type="any")$ID),1,0) 
  names(mpra)[ncol(mpra)]<-h
  }
  
# Format cage data
  # Make GRange for CAGE data
  cage.se<-makeGRangesFromDataFrame(cage, seqnames = "V1", start.field = "V2", end.field = "V3", keep.extra.columns = F) 
  # Add in new column in MPRA file with whether or not it is a CAGE-based enhancer
  mpra$cage<-ifelse(mpra$ID%in%(subsetByOverlaps( mpra.se, cage.se,type="any")$ID),1,0) 
  
# Intergrate cage, histone and mpra data
  mpra$dhs<-mpra$dhs_Tcell_merged #Get merged T cell DHS column
  dat.enrichment<-data.frame(mark=rep(c(histone.list, "cage", "dhs"), each=2), mode=rep(c("pcre", "emvar"), times=8), a=0, b=0, c=0, d=0, fold=0, p=0, odds=0, upper.conf=0, lower.conf=0, stringsAsFactors = F)

########### Now for the rest of the analysis  
  
# Calculate p-value and enrichment values
  for(i in 1:nrow(dat.enrichment)){
  cell_col<-which(names(mpra)==dat.enrichment[i,]$mark) #Get column of MPRA file corresponding to current anntotation
  if(dat.enrichment[i,]$mode=="emvar"){ #Run enrichment for emVars in the annotation
    a<-nrow(mpra[mpra$mpra_sig=="Enhancer_Skew" & mpra[,cell_col]==1,])
    b<-nrow(mpra[mpra$mpra_sig=="Enhancer_Skew"& mpra[,cell_col]==0,])
    c<-nrow(mpra[mpra$mpra_sig!="Enhancer_Skew" & mpra[,cell_col]==1,])
    d<-nrow(mpra[mpra$mpra_sig!="Enhancer_Skew" & mpra[,cell_col]==0,])
  }
  if(dat.enrichment[i,]$mode=="pcre"){ #Run enrichment for pCREs in the annotation
    a<-nrow(mpra[mpra$mpra_sig%in%c("Enhancer_Skew", "Enhancer_nSkew") & mpra[,cell_col]==1,])
    b<-nrow(mpra[mpra$mpra_sig%in%c("Enhancer_Skew", "Enhancer_nSkew")& mpra[,cell_col]==0,])
    c<-nrow(mpra[mpra$mpra_sig=="nEnhancer_nSkew" & mpra[,cell_col]==1,])
    d<-nrow(mpra[mpra$mpra_sig=="nEnhancer_nSkew" & mpra[,cell_col]==0,])
  }
  dat.enrichment[i,]$a<-a
  dat.enrichment[i,]$b<-b
  dat.enrichment[i,]$c<-c
  dat.enrichment[i,]$d<-d
  dat.enrichment[i,]$fold<-(a/(a+c))/(b/(b+d)) #Calculate enrichment
  dat.enrichment[i,]$p<-fisher.test(rbind(c(a,b), c(c, d)), alternative="two.sided")$p.value #Calculate p-value
  dat.enrichment[i,]$odds<-fisher.test(rbind(c(a,b), c(c, d)))$estimate #Calculate odds ratio
  dat.enrichment[i,]$lower.conf<-fisher.test(rbind(c(a,b), c(c, d)))$conf.int[1] #Calculate lower confidence interval
  dat.enrichment[i,]$upper.conf<-fisher.test(rbind(c(a,b), c(c, d)))$conf.int[2] #Calculate higher confidence interval
  }
# Make variant category and histone marks a factor in the analysis
dat.enrichment$mode<-factor(dat.enrichment$mode, levels=c("pcre", "emvar"))
dat.enrichment$mark<-factor(dat.enrichment$mark, levels=c("H3K4me1","H3K4me3", "H3K9me3", "H3K27ac","H3K27me3", "H3K36me3", "cage", "dhs"))
dat.enrichment$significant<-ifelse(dat.enrichment$p<(0.05/8), "*", "")

ylimit <- max(dat.enrichment$fold) + 1

write.table(dat.enrichment, paste0(data.dir,"/", mpra.name1, "_histone_cage_dhs_data.txt"), row.names=FALSE,col.names=TRUE,sep="\t", quote=FALSE)

# Plot proportion of MPRA variants overlapping chromHMM, DHS, or CAGE
  histone_marks_plot <- ggplot(dat.enrichment, aes(fill=mode, y=fold, x=mark)) + 
  geom_bar(position="dodge", stat="identity", color="black")+
  theme_bw()+scale_fill_brewer(palette = "Paired")+coord_cartesian(ylim=c(0,ylimit))+
  theme(axis.text.x = element_text( angle = 45, hjust=1))+ 
  geom_hline(yintercept=1, linetype="dashed", color = "black") + ggtitle("Fold change in pCREs and emVars in histone marks, cage and dhs", subtitle =mpra1_name) + 
    geom_text(aes(label=signif(p,2), group=mode),position = position_dodge(width=0.9), vjust=-1.5) # + geom_text(aes(label=significant), position=position_dodge(width=0.9), vjust=-0.25) 
  histone_marks_plot

# Format chromhmm
  chromhmm<-chromhmm[,c(1:4)]
  names(chromhmm)<-c("chr", "start", "end", "chromhmm")
  #Add in membership of chromHMM segmentations for MPRA variants
  mpra$chromhmm<-NA
  for(i in 1:nrow(mpra)){
    mpra[i,]$chromhmm<-subset(chromhmm, chr==mpra[i,]$chr & start<mpra[i,]$snp_end & end>=mpra[i,]$snp_end)$chromhmm[1]
    }

  #Calculate enrichments of pCREs and emVars in chromHMM segmentations
  dat.enrichment<-data.frame(chromhmm=rep(unique(chromhmm$chromhmm), 2),mode=rep(c("pcre", "emvar"), each=length(unique(chromhmm$chromhmm))), a=0, b=0, c=0, d=0,fold=0, p=0, odds=0, upper.conf=0, lower.conf=0, stringsAsFactors = F)
  for(i in 1:nrow(dat.enrichment)){
  if(dat.enrichment[i,]$mode=="pcre"){ #Run enrichment for pCREs in the chromHMM segmentation
    a<-nrow(subset(mpra, mpra_sig!="nEnhancer_nSkew" & chromhmm==dat.enrichment[i,]$chromhmm))
    b<-nrow(subset(mpra, mpra_sig!="nEnhancer_nSkew" & chromhmm=="18_Quies"))
    c<-nrow(subset(mpra, mpra_sig=="nEnhancer_nSkew" & chromhmm==dat.enrichment[i,]$chromhmm))
    d<-nrow(subset(mpra, mpra_sig=="nEnhancer_nSkew" & chromhmm=="18_Quies"))
  }
  if(dat.enrichment[i,]$mode=="emvar"){ #Run enrichment for emVars in the chromHMM segmentation
    a<-nrow(subset(mpra, mpra_sig=="Enhancer_Skew" & chromhmm==dat.enrichment[i,]$chromhmm))
    b<-nrow(subset(mpra, mpra_sig=="Enhancer_Skew" & chromhmm=="18_Quies"))
    c<-nrow(subset(mpra, mpra_sig=="nEnhancer_nSkew" & chromhmm==dat.enrichment[i,]$chromhmm))
    d<-nrow(subset(mpra, mpra_sig=="nEnhancer_nSkew" & chromhmm=="18_Quies"))
  }
  dat.enrichment[i,]$a<-a
  dat.enrichment[i,]$b<-b
  dat.enrichment[i,]$c<-c
  dat.enrichment[i,]$d<-d
  dat.enrichment[i,]$fold<-(a/(a+b))/(c/(c+d)) #Calculate enrichment
  dat.enrichment[i,]$p<-fisher.test(rbind(c(a,b), c(c, d)), alternative="two.sided")$p.value #calculate p-value
  dat.enrichment[i,]$odds<-fisher.test(rbind(c(a,b), c(c, d)))$estimate #Calculate odds ratio
  dat.enrichment[i,]$lower.conf<-fisher.test(rbind(c(a,b), c(c, d)))$conf.int[1] #Calculate lower confidence interval
  dat.enrichment[i,]$upper.conf<-fisher.test(rbind(c(a,b), c(c, d)))$conf.int[2] #Calculate higher confidence interval
  }

# Format enrichment data for plot  
  dat.enrichment$segmentation<-as.numeric(str_split_fixed(dat.enrichment$chromhmm, "\\_", 2)[,1])
  dat.enrichment<-dat.enrichment[order(dat.enrichment$segmentation),]
  dat.enrichment$chromhmm<-factor(dat.enrichment$chromhmm, levels=unique(dat.enrichment$chromhmm))
  dat.enrichment$mode<-factor(dat.enrichment$mode, levels=c("pcre", "emvar"))
  dat.enrichment$significant<-ifelse(dat.enrichment$p<(0.05/36), "***", "")
  dat.enrichment[dat.enrichment$p<0.05 & dat.enrichment$p>=0.05/36,]$significant<-"*"

ylimit <- max(dat.enrichment$fold) + 5
  
write.table(dat.enrichment, paste0(data.dir,"/", mpra.name1, "_chrommhmm_histone_data.txt"), row.names=FALSE,col.names=TRUE,sep="\t", quote=FALSE)

# plot of enrichment of chromatin marks
chromatin_marks_plot <- ggplot(dat.enrichment, aes(fill=mode, y=fold, x=chromhmm)) + 
  geom_bar(position="dodge", stat="identity", color="black")+
  theme_bw()+scale_fill_brewer(palette = "Paired")+coord_cartesian(ylim=c(0,ylimit))+
  theme(axis.text.x = element_text( angle = 45, hjust=1))+ 
  geom_hline(yintercept=1, linetype="dashed", color = "black") + ggtitle("Fold changes between pCREs and emVars in chromatin marks", subtitle=mpra1_name) +
  geom_text(data=subset(dat.enrichment, fold>3) ,aes(label=signif(p,2), group=mode),position = position_dodge(width=0.9), vjust=-1.5)
# + geom_text(aes(label=significant), position=position_dodge(width=0.9), vjust=-0.25)
chromatin_marks_plot



  # creating the plot name from the mpra.name1
plots_name1 <- paste0(mpra.name1,"_histone_marks",".pdf")
# Saving the plot in the plots directory 
ggsave(plots_name1,plot = histone_marks_plot, path = plots.dir, width = 7, height = 7)

# creating the plot name from the mpra.name1
plots_name2 <- paste0(mpra.name1,"_chromatin_marks",".pdf")
# Saving the plot in the plots directory 
ggsave(plots_name2,plot = chromatin_marks_plot, path = plots.dir, width = 7, height = 7)
```


# Motifbreakr function
```{r motifbeakr table, eval=FALSE}
# I am putting this in the form of a function because it takes a very long time to run. Also it is going to be eval=FALSE so the HTML doesn't take forever to be created. 
mpra_motifbreakr1 <- function(mpra,motifbreakr.bed.dir,motifbreakr.results.dir){
# subset mpra to only columns you need
mpra<-subset(mpra,project=="TGWAS" , select=c(chr, snp_start, snp_end, SNP, mpra_sig, LogSkew))

mpra$name<-paste(mpra$chr, mpra$snp_end, str_split_fixed(mpra$SNP, "\\:", 4)[,3], str_split_fixed(mpra$SNP, "\\:", 4)[,4],sep=":")
# Making a score column of all zeros
mpra$score<-0
# Add strand column
mpra$strand<-"+"
# subset mpra.bed to only the columns you want
mpra.bed<-subset(mpra, select=c(chr, snp_start, snp_end, name,score, strand))
write.table(mpra.bed, paste0(motifbreakr.bed.dir), row.names = F, col.names = F, sep="\t", quote=F)

#import the BED file
snps.mb.frombed <- snps.from.file(file = paste0(motifbreakr.bed.dir),dbSNP = SNPlocs.Hsapiens.dbSNP144.GRCh37,search.genome = BSgenome.Hsapiens.UCSC.hg19,format = "bed")

#Run motifbreakr and write results out
data("hocomoco")
hocomoco
# Run the motifbreakr function
results <- motifbreakR(snpList =snps.mb.frombed, filterp = TRUE,
                       pwmList = hocomoco,
                       threshold = 1e-5,
                       method = "log")
dat.results<-data.frame(results, stringsAsFactors = F)
dat.results <- apply(dat.results,2,as.character)
# Create mpra.motifbreakr.method_log.p1e5.results.txt
write.table(dat.results, paste0(motifbreakr.results.dir), row.names = F, col.names = T, quote=F, sep="\t")

return(dat.results)
}

# mpra (must have these columns: chr, snp_start, snp_end, SNP, mpra_sig, LogSkew) 
mpra<- mpra_merge
# motifbreakr.bed.dir 
motifbreakr.bed.dir <- paste0(data.dir, "tf/",mpra.name1,"_hg19_mpra_motifbreakr.bed")
# motifbreakr.results.dir
motifbreakr.results.dir <- paste0(data.dir,"tf/",mpra.name1,"_motifbreakr_results.txt")

mpra_motifbreakr1(mpra=mpra, motifbreakr.bed.dir=motifbreakr.bed.dir,motifbreakr.results.dir=motifbreakr.results.dir)
```

# Motifbreakr plots
```{r motifbreakr plots}
# dat.tf (Read in TF ChIP binding sites from ENCODE)
dat.tf<-fread("https://hgdownload.cse.ucsc.edu/goldenpath/hg19/encodeDCC/wgEncodeRegTfbsClustered/wgEncodeRegTfbsClusteredV3.bed.gz")
# Where would you like the motifbreakr bed file to be located and what would you like it to be called? Please end file name in .bed 
motifbreakr.bed.dir 
# # Where would you like the motifbreakr results file to be located and what would you like it to be called? Please end file name in .txt
motifbreakr.results.dir

#####Evaluate overlap of MPRA variants with motifbreakR#######

  mpra <- mpra_merge
  
# Read in main MPRA table and subset for the column we want
  mpra<-subset(mpra, project=="TGWAS", select=c( chr, snp_start, snp_end, SNP,A.log2FC,B.log2FC,LogSkew,mpra_sig))
  # Make mpra into a Grandes object
  mpra.se<-makeGRangesFromDataFrame(mpra, seqnames = "chr", start.field = "snp_start", end.field = "snp_end",keep.extra.columns=TRUE)

# Read in Motifbreakr data and merge with MPRA
  dat.motifbreakr<-read.delim(paste0(motifbreakr.results.dir),header=T, stringsAsFactors = F,sep="\t")
  # Subset motifbreakr data into only the columns we want 
  dat.motifbreakr<-subset(dat.motifbreakr,  select=c(seqnames, end, REF, ALT,SNP_id,  geneSymbol,scoreRef ,scoreAlt, alleleDiff))
  # Making a SNP columns in the motifbreakr data
  dat.motifbreakr$SNP<-gsub("chr", "", dat.motifbreakr$SNP_id)
  # merge mpra and motifbreakr data
  dat.motifbreakr<-merge(dat.motifbreakr, subset(mpra, select=c(SNP, A.log2FC,B.log2FC,LogSkew,mpra_sig)), by="SNP", all.x=T, all.y=F)

# Format TF ChIP binding sites from ENCODE
  names(dat.tf)[c(1:4,7)]<-c("chr", "start", "end", "tf", "cell")
  #subset ENCODE TF data for txn factors identified in the motifbreakR calls
  dat.tf<-subset(dat.tf, tf%in%dat.motifbreakr$geneSymbol, select=c(chr, start, end, tf, cell))
  # make tf data a grange object
  tf.se<-makeGRangesFromDataFrame(dat.tf, seqnames = "chr", start.field = "start", end.field = "end",keep.extra.columns=TRUE)

#Intersect MPRA variants with ENCODE TF data
  dat.tf.overlap<-data.frame(subsetByOverlaps(tf.se, mpra.se, type="any"), stringsAsFactors = F)
  #Find the cell types from ENCODE data where there is a TF binding site at each motifbreakR hit
  dat.motifbreakr$overlap<-0
for(i in 1:nrow(dat.motifbreakr)){
  if(nrow(subset(dat.tf.overlap, tf==dat.motifbreakr[i,]$geneSymbol & seqnames==dat.motifbreakr[i,]$seqnames & start<=dat.motifbreakr[i,]$end & end>=dat.motifbreakr[i,]$end   ))){
    dat.motifbreakr[i,]$overlap<-subset(dat.tf.overlap, tf==dat.motifbreakr[i,]$geneSymbol & seqnames==dat.motifbreakr[i,]$seqnames & start<=dat.motifbreakr[i,]$end & end>=dat.motifbreakr[i,]$end   )$cell[1]
  }  
}
  
# Generate unique MPRA sites
  dat.motifbreakr.unique<-subset(dat.motifbreakr,  overlap!=0)
  dat.motifbreakr.unique<-dat.motifbreakr.unique[order(-abs(dat.motifbreakr.unique$alleleDiff)),]
  dat.motifbreakr.unique<-dat.motifbreakr.unique[!duplicated(dat.motifbreakr.unique$SNP_id),]

# Get a list of SNPs from the bed file
  snps.mb.frombed <- snps.from.file(file = paste0(motifbreakr.bed.dir),dbSNP = SNPlocs.Hsapiens.dbSNP144.GRCh37,
                                    search.genome = BSgenome.Hsapiens.UCSC.hg19,format = "bed")
# Subset MPRA file for SNPs evaluated by motifbreakR
  snp.list<-snps.mb.frombed$SNP_id
  mpra<-mpra[paste0("chr", mpra$SNP)%in%snps.mb.frombed$SNP_id,]

# Plot proportions of variants that have allelic TF binding
# Produces Figure S4B
  dat.plot<-data.frame( mpra_sig=c("nEnhancer_nSkew", "Enhancer_nSkew", "Enhancer_Skew"), proportion=0,total=0,stringsAsFactors = F)
  for(i in 1:nrow(dat.plot)){
  dat.plot[i,]$proportion<-nrow(subset(dat.motifbreakr.unique, mpra_sig==dat.plot[i,]$mpra_sig))/nrow(subset(mpra, mpra_sig==dat.plot[i,]$mpra_sig))
  }
  for(i in 1:nrow(dat.plot)){
  dat.plot[i,]$total<-nrow(subset(dat.motifbreakr.unique, mpra_sig==dat.plot[i,]$mpra_sig))
  }
  dat.plot$mpra_sig<-factor(dat.plot$mpra_sig, levels=c("nEnhancer_nSkew", "Enhancer_nSkew", "Enhancer_Skew"))
  # Calculate y axis limit for proportion plot
  ylimit <- max(dat.plot$proportion) + 0.01
  
 #Test significance of overlap between TFs with allelic activity and MPRA variants
mpra<-mpra_merge

#Perform for pCREs
proportion_test_pcre <- prop.test(x = c(nrow(subset(dat.motifbreakr.unique , mpra_sig=="Enhancer_nSkew")), nrow(subset(dat.motifbreakr.unique, mpra_sig=="nEnhancer_nSkew"))), 
          n =c(nrow(subset(mpra, snp_end-snp_start==1 & mpra_sig=="Enhancer_nSkew")), nrow(subset(mpra, snp_end-snp_start==1 & mpra_sig=="nEnhancer_nSkew"))))
proportion_test_pcre
#Perform for emVars
proportion_test_emvar <- prop.test(x = c(nrow(subset(dat.motifbreakr.unique , mpra_sig=="Enhancer_Skew")), nrow(subset(dat.motifbreakr.unique, mpra_sig=="nEnhancer_nSkew"))), 
          n =c(nrow(subset(mpra, snp_end-snp_start==1 & mpra_sig=="Enhancer_Skew")), nrow(subset(mpra, snp_end-snp_start==1 & mpra_sig=="nEnhancer_nSkew")))) 
proportion_test_emvar
  dat.plot$p <- c(NA, proportion_test_pcre$p.value, proportion_test_emvar$p.value)

# Proportion plot of pCREs and emVars that have TF binding 
  motifbreakr_plot <- ggplot(dat.plot, aes(fill=mpra_sig, y=proportion, x=mpra_sig)) + 
    geom_bar(position="dodge", stat="identity", color="black")+
    theme_bw()+scale_fill_brewer(palette = "Blues")+coord_cartesian(ylim=c(0,ylimit))+
    theme(axis.text.x = element_text( angle = 45, hjust=1))+ 
    geom_hline(yintercept=1, linetype="dashed", color = "black") + 
    ggtitle("Proportion of pCREs and emVars that have TF allelic binding according to motifbreakr", subtitle = mpra1_name) + 
    xlab("Variant category") + 
    ylab("Proportion of variants with allelic TF binding") +
    geom_text(aes(label=signif(p,2)), position="dodge", vjust=-0.75) +
    geom_text(aes(label=round(total,3)), position="dodge", vjust=1.5) 
motifbreakr_plot

#Plot MPRA allelic skew against TF allelic difference from motifbreakR for emVars
dat.motifbreakr.emvar<-subset(dat.motifbreakr.unique, mpra_sig==c("Enhancer_Skew"))

allelic_skew_vs_tf_skew_plot <- plot(x=dat.motifbreakr.emvar$alleleDiff, y=dat.motifbreakr.emvar$LogSkew, pch=16, cex=1.2 ,col="forestgreen", xlim=c(-6,6), ylim=c(-1.5,1.5))
abline(v=0, lty=2, col="blue")
abline(h=0, lty=2, col="blue")
abline(lm(dat.motifbreakr.emvar$LogSkew~dat.motifbreakr.emvar$alleleDiff))

motifbreakr_mpra_corr_model <- summary(lm(dat.motifbreakr.emvar$LogSkew~dat.motifbreakr.emvar$alleleDiff))
motifbreakr_mpra_corr_model

mpra_logskew_tf_skew_ggplot <- ggplot(data=dat.motifbreakr.emvar, aes(x=alleleDiff, y=LogSkew)) + geom_point(color="#E14A4A") + theme_bw()  + geom_hline(yintercept=0,color="#5D93CD") +  geom_vline(xintercept=0,color="#5D93CD") + geom_smooth(method = "lm",color="#5D93CD",fill="#5D93CD") + xlim(-7,7) + ylim(-2,2) + labs(title="motifbreakr allelediff vs. mpra logskew correlation plot", subtitle=mpra1_name, caption= paste0("r^2: ", motifbreakr_mpra_corr_model$r.squared, "; p-value: ", motifbreakr_mpra_corr_model$coefficients[2,4]))
print(mpra_logskew_tf_skew_ggplot)

if(plots.dir != "NA"){
# creating the plot name from the mpra.name1
plots_name <- paste0(mpra.name1,"_motifbreakr_mpra_correlation",".pdf")
# Saving the plot in the plots directory 
ggsave(plots_name,plot = mpra_logskew_tf_skew_ggplot, path = plots.dir, width = 7, height = 7)
}

#Calculate statistical significance for correlation between MPRA alleliic skew against TF allelic difference
skew_vs_skew_model <- summary(lm(dat.motifbreakr.emvar$LogSkew~dat.motifbreakr.emvar$alleleDiff))
skew_vs_skew_model

# creating the plot name from the mpra.name1
plots_name <- paste0(mpra.name1,"_motifbreakr",".pdf")
# Saving the plot in the plots directory 
ggsave(plots_name,plot = motifbreakr_plot, path = plots.dir, width = 7, height = 7)
```





